// motia/dist/cjs/__tests__/openapi/generate.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const generate_1 = require("../../openapi/generate");
jest.mock('fs', () => ({
    readFileSync: jest.fn(),
    writeFileSync: jest.fn(),
}));
describe('generateOpenApi', () => {
    const mockProjectDir = '/mock/project';
    const mockOutputFile = 'mock-project-openapi.json';
    beforeEach(() => {
        jest.clearAllMocks();
        fs.readFileSync.mockReturnValue(JSON.stringify({ name: 'test-project', version: '1.0.0' }));
    });
    const MOCK_API_STEP = {
        filePath: '/mock/project/steps/test.ts',
        version: '1.0.0',
        config: {
            type: 'api',
            name: 'testApi',
            path: '/test',
            method: 'GET',
            emits: [],
        },
    };
    it('should generate a basic OpenAPI spec with default info', () => {
        const apiSteps = [MOCK_API_STEP];
        (0, generate_1.generateOpenApi)(mockProjectDir, apiSteps, undefined, undefined, mockOutputFile);
        const expectedOpenApi = {
            openapi: '3.0.0',
            info: {
                title: 'test-project',
                version: '1.0.0',
            },
            paths: {
                '/test': {
                    get: {
                        summary: 'testApi',
                        description: undefined,
                        requestBody: undefined,
                        responses: {},
                    },
                },
            },
            components: {
                schemas: {},
            },
        };
        expect(fs.writeFileSync).toHaveBeenCalledWith(path.join(mockProjectDir, mockOutputFile), JSON.stringify(expectedOpenApi, null, 2));
    });
    it('should handle bodySchema and responseSchema correctly', () => {
        const mockBodySchema = {
            type: 'object',
            properties: {
                id: { type: 'string' },
            },
        };
        const mockResponseSchema = {
            200: {
                type: 'object',
                properties: {
                    message: { type: 'string' },
                },
            },
        };
        const mockApiStep = {
            ...MOCK_API_STEP,
            config: {
                ...MOCK_API_STEP.config,
                type: 'api',
                name: 'testApiWithSchemas',
                path: '/test-schemas',
                method: 'POST',
                bodySchema: mockBodySchema,
                responseSchema: mockResponseSchema,
                emits: [],
            },
        };
        const apiSteps = [mockApiStep];
        (0, generate_1.generateOpenApi)(mockProjectDir, apiSteps, undefined, undefined, mockOutputFile);
        const expectedOpenApi = {
            openapi: '3.0.0',
            info: {
                title: 'test-project',
                version: '1.0.0',
            },
            paths: {
                '/test-schemas': {
                    post: {
                        summary: 'testApiWithSchemas',
                        description: undefined,
                        requestBody: {
                            content: {
                                'application/json': {
                                    schema: mockBodySchema,
                                },
                            },
                        },
                        responses: {
                            200: {
                                description: 'Response for status code 200',
                                content: {
                                    'application/json': {
                                        schema: mockResponseSchema[200],
                                    },
                                },
                            },
                        },
                    },
                },
            },
            components: {
                schemas: {},
            },
        };
        expect(fs.writeFileSync).toHaveBeenCalledWith(path.join(mockProjectDir, mockOutputFile), JSON.stringify(expectedOpenApi, null, 2));
    });
    it('should use provided title', () => {
        const apiSteps = [MOCK_API_STEP];
        const customTitle = 'Custom API';
        (0, generate_1.generateOpenApi)(mockProjectDir, apiSteps, customTitle);
        const expectedOpenApi = {
            openapi: '3.0.0',
            info: {
                title: customTitle,
                version: '1.0.0',
            },
            paths: {
                '/test': {
                    get: {
                        summary: 'testApi',
                        description: undefined,
                        requestBody: undefined,
                        responses: {},
                    },
                },
            },
            components: {
                schemas: {},
            },
        };
        expect(fs.writeFileSync).toHaveBeenCalledWith(path.join(mockProjectDir, 'openapi.json'), JSON.stringify(expectedOpenApi, null, 2));
    });
    it('should use provided version', () => {
        const apiSteps = [MOCK_API_STEP];
        const customVersion = '2.0.0';
        (0, generate_1.generateOpenApi)(mockProjectDir, apiSteps, undefined, customVersion, mockOutputFile);
        const expectedOpenApi = {
            openapi: '3.0.0',
            info: {
                title: 'test-project',
                version: customVersion,
            },
            paths: {
                '/test': {
                    get: {
                        summary: 'testApi',
                        description: undefined,
                        requestBody: undefined,
                        responses: {},
                    },
                },
            },
            components: {
                schemas: {},
            },
        };
        expect(fs.writeFileSync).toHaveBeenCalledWith(path.join(mockProjectDir, mockOutputFile), JSON.stringify(expectedOpenApi, null, 2));
    });
    it('should handle complex bodySchema with anyOf, nullable, and $ref', () => {
        const mockUserSchema = {
            type: 'object',
            properties: {
                name: { type: 'string' },
                email: { type: 'string', format: 'email' },
            },
            required: ['name', 'email'],
        };
        const mockComplexBodySchema = {
            type: 'object',
            properties: {
                id: { type: 'string' },
                status: {
                    anyOf: [{ type: 'string', enum: ['active', 'inactive'] }, { type: 'number', enum: [0, 1] }, { type: 'null' }],
                },
                user: { $ref: '#/components/schemas/User' },
            },
            required: ['id'],
            $defs: {
                User: mockUserSchema,
            },
        };
        const mockApiStep = {
            ...MOCK_API_STEP,
            filePath: '/mock/project/steps/complex.ts',
            config: {
                ...MOCK_API_STEP.config,
                type: 'api',
                name: 'testComplexApi',
                path: '/test-complex',
                method: 'PUT',
                emits: [],
                bodySchema: mockComplexBodySchema,
            },
        };
        const apiSteps = [mockApiStep];
        (0, generate_1.generateOpenApi)(mockProjectDir, apiSteps, undefined, undefined, mockOutputFile);
        const expectedOpenApi = {
            openapi: '3.0.0',
            info: {
                title: 'test-project',
                version: '1.0.0',
            },
            paths: {
                '/test-complex': {
                    put: {
                        summary: 'testComplexApi',
                        description: undefined,
                        requestBody: {
                            content: {
                                'application/json': {
                                    schema: {
                                        type: 'object',
                                        properties: {
                                            id: { type: 'string' },
                                            status: {
                                                anyOf: [
                                                    { type: 'string', enum: ['active', 'inactive'] },
                                                    { type: 'number', enum: [0, 1] },
                                                ],
                                                nullable: true,
                                            },
                                            user: { $ref: '#/components/schemas/User' },
                                        },
                                        required: ['id'],
                                    },
                                },
                            },
                        },
                        responses: {},
                    },
                },
            },
            components: {
                schemas: {
                    User: mockUserSchema,
                },
            },
        };
        expect(fs.writeFileSync).toHaveBeenCalledWith(path.join(mockProjectDir, mockOutputFile), JSON.stringify(expectedOpenApi, null, 2));
    });
});
