// motia/dist/cjs/cloud/new-deployment/listeners/build-printer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildPrinter = void 0;
const printer_1 = require("@motiadev/core/dist/src/printer");
const colors_1 = __importDefault(require("colors"));
const cli_output_manager_1 = require("../../cli-output-manager");
const pretty_bytes_1 = require("../utils/pretty-bytes");
const building = colors_1.default.yellow('➜ [BUILDING]');
const built = colors_1.default.green('✓ [BUILT]');
const failed = colors_1.default.red('✘ [FAILED]');
const skipped = colors_1.default.gray('- [SKIPPED]');
const baseTag = (tag) => colors_1.default.bold(colors_1.default.magenta(tag));
const streamTag = baseTag('Stream');
const stepTag = baseTag('Step');
const routerTag = baseTag('Router');
class BuildPrinter {
    constructor() {
        this.printer = new printer_1.Printer(process.cwd());
        this.output = new cli_output_manager_1.CLIOutputManager();
    }
    getLanguage(language) {
        if (language === 'python') {
            return colors_1.default.bold(colors_1.default.blue('Python'));
        }
        else if (language === 'node') {
            return colors_1.default.bold(colors_1.default.green('Node'));
        }
        else if (language === 'ruby') {
            return colors_1.default.bold(colors_1.default.red('Ruby'));
        }
        return colors_1.default.bold(colors_1.default.gray('Unknown'));
    }
    getStepLanguage(step) {
        if (step.filePath.endsWith('.py')) {
            return this.getLanguage('python');
        }
        else if (step.filePath.endsWith('.js') || step.filePath.endsWith('.ts')) {
            return this.getLanguage('node');
        }
        else if (step.filePath.endsWith('.rb')) {
            return this.getLanguage('ruby');
        }
        return this.getLanguage('unknown');
    }
    printStepBuilding(step, progressMessage) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => {
            message.append(`${building} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`);
            if (progressMessage) {
                message.append(colors_1.default.yellow(progressMessage));
            }
        });
    }
    printStepBuilt(step, size) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message
            .append(`${built} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`)
            .append(`${(0, pretty_bytes_1.prettyBytes)(size)}`, 'gray'));
    }
    printApiRouterBuilding(language) {
        const fileName = `router-${language}.zip`;
        const coloredFileName = colors_1.default.bold(colors_1.default.cyan(fileName));
        this.output.log(fileName, (message) => message //
            .append(routerTag)
            .append(building)
            .append(this.getLanguage(language))
            .append(coloredFileName));
    }
    printApiRouterBuilt(language, size) {
        const fileName = `router-${language}.zip`;
        const coloredFileName = colors_1.default.bold(colors_1.default.cyan(fileName));
        this.output.log(fileName, (message) => message
            .append(built)
            .append(routerTag)
            .append(this.getLanguage(language))
            .append(coloredFileName)
            .append((0, pretty_bytes_1.prettyBytes)(size), 'gray'));
    }
    printStepFailed(step, error) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message.append(`${failed} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`).append(error.message, 'red'));
    }
    printStepSkipped(step, reason) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message.append(`${skipped} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`).append(reason, 'yellow'));
    }
    printStreamCreated(stream) {
        const streamPath = this.getStreamPath(stream);
        const streamName = colors_1.default.gray(`[${stream.config.name}]`);
        this.output.log(stream.filePath, (message) => message.append(`${built} ${streamTag} ${streamName} ${streamPath}`));
    }
    getStreamPath(stream) {
        return colors_1.default.bold(colors_1.default.cyan(this.printer.getRelativePath(stream.filePath)));
    }
}
exports.BuildPrinter = BuildPrinter;
