// motia/dist/cjs/cloud/new-deployment/listeners/cli-listener.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliListener = void 0;
const colors_1 = __importDefault(require("colors"));
const build_printer_1 = require("./build-printer");
const print_deployment_status_1 = require("./print-deployment-status");
const printer_1 = require("./printer");
class CliListener {
    constructor(context) {
        this.context = context;
        this.deployPrinter = new printer_1.DeployPrinter();
        this.printer = new build_printer_1.BuildPrinter();
    }
    onBuildStart(step) {
        this.printer.printStepBuilding(step);
    }
    onBuildProgress(step, message) {
        this.printer.printStepBuilding(step, message);
    }
    onBuildEnd(step, size) {
        this.printer.printStepBuilt(step, size);
    }
    onBuildError(step, error) {
        this.printer.printStepFailed(step, error);
    }
    onBuildSkip(step, reason) {
        this.printer.printStepSkipped(step, reason);
    }
    onStreamCreated(stream) {
        this.printer.printStreamCreated(stream);
    }
    onApiRouterBuilding(language) {
        this.printer.printApiRouterBuilding(language);
    }
    onApiRouterBuilt(language, size) {
        this.printer.printApiRouterBuilt(language, size);
    }
    onWarning(id, warning) {
        this.context.log(id, (message) => message.tag('warning').append(warning));
    }
    onBuildWarning(warning) {
        this.context.log(`build-warnings-${warning.step.filePath}`, (message) => {
            message.tag('warning').append(warning.message);
        });
    }
    onBuildErrors(errors) {
        this.context.log('build-failed', (message) => {
            message.box(['Unable to deploy to Motia Cloud, please fix the following errors'], 'red');
        });
        errors.map((error) => {
            const filePath = colors_1.default.gray(`[${error.relativePath}]`);
            this.context.log(`build-errors-${error.relativePath}`, (message) => {
                message.tag('failed').append(`${filePath} ${error.message}`);
            });
        });
        this.context.log('build-failed-end', (message) => {
            message.append(colors_1.default.gray('\n--------------------------------\n'));
            message.tag('failed').append('Deployment canceled', 'red');
        });
    }
    stepUploadStart(_, step) {
        this.deployPrinter.printStepUploading(step);
    }
    stepUploadEnd(_, step) {
        this.deployPrinter.printStepUploaded(step);
    }
    routeUploadStart(path, language) {
        this.deployPrinter.printRouterUploading(path, language);
    }
    routeUploadEnd(path, language) {
        this.deployPrinter.printRouterUploaded(path, language);
    }
    onDeployStart() {
        this.context.log('deploy-progress', (message) => message.tag('progress').append('Version in progress...'));
    }
    onDeployProgress(data) {
        (0, print_deployment_status_1.printDeploymentStatus)(data, this.context);
    }
    onDeployEnd({ output }) {
        this.context.log('deploy-result', (message) => message.tag('success').append('Deployment process completed successfully'));
        this.context.log('deploy-output', (message) => message.table(['Field', 'Value'], Object.entries(output).map(([key, value]) => [key, value])));
    }
    onDeployError(errorMessage) {
        this.context.log('deploy-result', (message) => message
            .tag('failed')
            .append('Deployment failed. Please check the deployment status and try again or contact support', 'red')
            .box([errorMessage], 'red'));
    }
    stepUploadProgress() { }
    routeUploadProgress() { }
    routeUploadError() { }
    stepUploadError() { }
}
exports.CliListener = CliListener;
