// motia/dist/cjs/cloud/new-deployment/listeners/printer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployPrinter = void 0;
const printer_1 = require("@motiadev/core/dist/src/printer");
const colors_1 = __importDefault(require("colors"));
const cli_output_manager_1 = require("../../cli-output-manager");
const constants_1 = require("../constants");
const uploading = colors_1.default.yellow('➜ [UPLOADING]');
const uploaded = colors_1.default.green('✓ [UPLOADED]');
class DeployPrinter {
    constructor() {
        this.printer = new printer_1.Printer(constants_1.projectDir);
        this.output = new cli_output_manager_1.CLIOutputManager();
    }
    getLanguage(language) {
        if (language === 'node') {
            return colors_1.default.bold(colors_1.default.green('Node'));
        }
        else if (language === 'python') {
            return colors_1.default.bold(colors_1.default.blue('Python'));
        }
        else if (language === 'ruby') {
            return colors_1.default.bold(colors_1.default.red('Ruby'));
        }
        return colors_1.default.bold(colors_1.default.gray('Unknown'));
    }
    getStepLanguage(stepConfig) {
        if (stepConfig.filePath.endsWith('.py')) {
            return this.getLanguage('python');
        }
        else if (stepConfig.filePath.endsWith('.js') || stepConfig.filePath.endsWith('.ts')) {
            return this.getLanguage('node');
        }
        else if (stepConfig.filePath.endsWith('.rb')) {
            return this.getLanguage('ruby');
        }
        return this.getLanguage('unknown');
    }
    getStepType(stepConfig) {
        if (stepConfig.config.type === 'api') {
            return colors_1.default.gray('(API)');
        }
        else if (stepConfig.config.type === 'event') {
            return colors_1.default.gray('(Event)');
        }
        else if (stepConfig.config.type === 'cron') {
            return colors_1.default.gray('(Cron)');
        }
        else if (stepConfig.config.type === 'noop') {
            return colors_1.default.gray('(Noop)');
        }
        return colors_1.default.gray('(Unknown)');
    }
    getRelativePath(path) {
        return colors_1.default.bold(colors_1.default.cyan(this.printer.getRelativePath(path)));
    }
    getStepPath(stepConfig) {
        return this.getRelativePath(stepConfig.filePath);
    }
    printStepUploading(stepConfig) {
        const stepLanguage = this.getStepLanguage(stepConfig);
        const stepType = this.getStepType(stepConfig);
        const stepPath = this.getStepPath(stepConfig);
        const stepTag = this.printer.stepTag;
        this.output.log(stepConfig.filePath, (message) => {
            message.append(`${uploading} ${stepLanguage} ${stepTag} ${stepType} ${stepPath}`);
        });
    }
    printStepUploaded(stepConfig) {
        const stepLanguage = this.getStepLanguage(stepConfig);
        const stepType = this.getStepType(stepConfig);
        const stepPath = this.getStepPath(stepConfig);
        const stepTag = this.printer.stepTag;
        this.output.log(stepConfig.filePath, (message) => {
            message.append(`${uploaded} ${stepLanguage} ${stepTag} ${stepType} ${stepPath}`);
        });
    }
    printRouterUploading(language, routerPath) {
        const stepLanguage = this.getLanguage(language);
        const stepPath = this.getRelativePath(routerPath);
        this.output.log(routerPath, (message) => message.append(`${uploading} ${stepLanguage} ${stepPath}`));
    }
    printRouterUploaded(language, routerPath) {
        const stepLanguage = this.getLanguage(language);
        const stepPath = this.getRelativePath(routerPath);
        this.output.log(routerPath, (message) => message.append(`${uploaded} ${stepLanguage} ${stepPath}`));
    }
    printConfigurationUploading() {
        this.output.log('upload-config', (message) => message.tag('progress').append(`${uploading} Configuration`));
    }
    printConfigurationUploaded() {
        this.output.log('upload-config', (message) => message.tag('success').append(`${uploaded} Configuration`));
    }
}
exports.DeployPrinter = DeployPrinter;
