// @motiadev/workbench/dist/src/publicComponents/base-node/code-display.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useThemeStore } from '@motiadev/ui';
import { useRef, useState } from 'react';
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { oneDark, oneLight } from 'react-syntax-highlighter/dist/esm/styles/prism';
import { FeatureCard } from './feature-card';
import { LanguageIndicator } from './language-indicator';
const codeTagProps = {
    style: {
        fontFamily: 'DM Mono, monospace',
        fontSize: '16px',
    },
};
const customStyle = {
    margin: 0,
    borderRadius: 0,
    padding: 0,
};
const isHighlighted = (lines, lineNumber) => {
    return lines.some((line) => {
        const [start, end] = line.split('-').map((num) => parseInt(num, 10));
        if (end !== undefined) {
            return lineNumber >= start && lineNumber <= end;
        }
        return lineNumber == start;
    });
};
const getFirstLineNumber = (line) => {
    const [start] = line.split('-').map((num) => parseInt(num, 10));
    return start;
};
export const CodeDisplay = ({ code, language, features }) => {
    const theme = useThemeStore((state) => state.theme);
    const themeStyle = theme === 'dark' ? oneDark : oneLight;
    const [highlightedLines, setHighlightedLines] = useState([]);
    const [selectedFeature, setSelectedFeature] = useState(null);
    const ref = useRef(null);
    const handleFeatureClick = (feature) => {
        setSelectedFeature(feature);
        setHighlightedLines(feature.lines);
        const lineNumber = getFirstLineNumber(feature.lines[0]);
        const line = ref.current?.querySelector(`[data-line-number="${lineNumber}"]`);
        if (line) {
            line.scrollIntoView({ behavior: 'smooth', block: 'center' });
        }
    };
    return (_jsxs("div", { className: "flex flex-col h-full overflow-hidden", children: [_jsxs("div", { className: "flex items-center py-2 px-5 dark:bg-[#1e1e1e] gap-2 justify-center", children: [_jsx("div", { className: "text-sm text-muted-foreground", children: "Read only" }), _jsx("div", { className: "flex-1" }), _jsx(LanguageIndicator, { language: language, className: "w-4 h-4", size: 16, showLabel: true })] }), _jsxs("div", { className: "flex flex-row h-[calc(100%-36px)]", children: [features && features.length > 0 && (_jsx("div", { className: "flex flex-col gap-2 p-2 bg-card overflow-y-auto min-w-[200px] w-[300px]", children: features.map((feature, index) => (_jsx(FeatureCard, { feature: feature, highlighted: selectedFeature === feature, onClick: () => handleFeatureClick(feature), onHover: () => handleFeatureClick(feature) }, index))) })), _jsx("div", { className: "overflow-y-auto", ref: ref, children: _jsx(SyntaxHighlighter, { showLineNumbers: true, language: language, style: themeStyle, codeTagProps: codeTagProps, customStyle: customStyle, wrapLines: true, lineProps: (lineNumber) => {
                                if (isHighlighted(highlightedLines, lineNumber)) {
                                    return {
                                        'data-line-number': lineNumber,
                                        style: {
                                            borderLeft: '2px solid var(--accent-1000)',
                                            backgroundColor: 'rgb(from var(--accent-1000) r g b / 0.2)',
                                        },
                                    };
                                }
                                return {
                                    'data-line-number': lineNumber,
                                    style: {
                                        borderLeft: '2px solid transparent',
                                    },
                                };
                            }, children: code }) })] })] }));
};
