// @motiadev/workbench/dist/src/publicComponents/base-node/node-header.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { cn } from '@motiadev/ui';
import { cva } from 'class-variance-authority';
import { CalendarClock, CircleOff, Link2, Waypoints } from 'lucide-react';
const baseIcon = cva('rounded-md p-2', {
    variants: {
        variant: {
            event: 'bg-[rgba(30,118,231,0.2)] text-[rgb(30,118,231)]',
            api: 'dark:bg-[rgba(192,255,17,0.2)] dark:text-[rgb(192,255,17)] bg-[rgb(231,255,166)] text-[rgb(94,125,11)]',
            noop: 'bg-[rgba(239,18,229,0.2)] text-[rgb(239,18,229)]',
            cron: 'bg-[rgba(241,105,15,0.2)] text-[rgb(241,105,15)]',
        },
    },
});
const NodeIcon = ({ variant }) => {
    if (variant === 'cron') {
        return _jsx(CalendarClock, { className: "w-5 h-5" });
    }
    else if (variant === 'api') {
        return _jsx(Link2, { className: "w-5 h-5" });
    }
    else if (variant === 'noop') {
        return _jsx(CircleOff, { className: "w-5 h-5" });
    }
    else if (variant === 'event') {
        return _jsx(Waypoints, { className: "w-5 h-5" });
    }
    return _jsx("div", { className: cn(baseIcon({ variant })), children: "Icon" });
};
export const NodeHeader = ({ text, variant, children, className }) => (_jsxs("div", { className: cn('flex items-center gap-2 p-2', className), children: [_jsx("div", { className: baseIcon({ variant }), children: _jsx(NodeIcon, { variant: variant }) }), _jsxs("div", { className: "flex flex-1 justify-between items-start gap-4", children: [_jsx("div", { className: "flex flex-col", children: _jsx("div", { className: "text-sm font-semibold leading-[1.25] tracking-[-0.25px]", children: text }) }), children] })] }));
