// @motiadev/workbench/dist/motia-plugin/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateImports = generateImports;
exports.generatePackageMap = generatePackageMap;
exports.generatePluginLogic = generatePluginLogic;
exports.generatePluginCode = generatePluginCode;
exports.generateCssImports = generateCssImports;
exports.isValidCode = isValidCode;
const resolver_1 = require("./resolver");
/**
 * Generates import statements for all unique plugin packages.
 *
 * @param packages - Array of unique package names
 * @returns JavaScript code string with import statements
 *
 * @example
 * ```ts
 * generateImports(['@org/plugin-1', '~/plugins/local'])
 * // Returns:
 * // import * as plugin_0 from '@org/plugin-1'
 * // import * as plugin_1 from '~/plugins/local'
 * ```
 */
function generateImports(packages) {
    return packages.map((packageName, index) => `import * as plugin_${index} from '${packageName}'`).join('\n');
}
/**
 * Generates the package map that links package names to their imported modules.
 *
 * @param packages - Array of unique package names
 * @returns JavaScript code string defining the package map
 *
 * @example
 * ```ts
 * generatePackageMap(['@org/plugin-1', '~/plugins/local'])
 * // Returns: const packageMap = {'@org/plugin-1': plugin_0,'~/plugins/local': plugin_1}
 * ```
 */
function generatePackageMap(packages) {
    const entries = packages.map((packageName, index) => `'${packageName}': plugin_${index}`);
    return `const packageMap = {${entries.join(',')}}`;
}
/**
 * Generates the plugin transformation logic that processes plugin configurations.
 *
 * @param plugins - Array of plugin configurations
 * @returns JavaScript code string with plugin processing logic
 */
function generatePluginLogic(plugins) {
    return `
      const motiaPlugins = ${JSON.stringify(plugins)}

      export const plugins = motiaPlugins.map((plugin) => {
        const component = packageMap[plugin.packageName]
        const config = component.config || {}
        const componentName = config.componentName || plugin.componentName

        return {
          label: plugin.label || config.label || 'Plugin label',
          labelIcon: plugin.labelIcon || config.labelIcon || 'toy-brick',
          position: plugin.position || config.position || 'top',
          props: plugin.props || config.props || {},
          component: componentName ? component[componentName] : component.default,
        }
      })
`;
}
/**
 * Generates the complete virtual module code for all plugins.
 * This is the main code generation function that combines all parts.
 *
 * @param plugins - Array of plugin configurations
 * @returns Complete JavaScript code string for the virtual module
 *
 * @example
 * ```ts
 * const plugins = [
 *   { packageName: '@test/plugin', label: 'Test' }
 * ]
 * const code = generatePluginCode(plugins)
 * // Returns complete module code with imports, map, and logic
 * ```
 */
function generatePluginCode(plugins) {
    if (!plugins || plugins.length === 0) {
        return 'export const plugins = []';
    }
    const packages = (0, resolver_1.getUniquePackageNames)(plugins);
    const imports = generateImports(packages);
    const packageMap = generatePackageMap(packages);
    const logic = generatePluginLogic(plugins);
    return `${imports}
${packageMap}
${logic}`;
}
/**
 * Generates CSS imports for plugins that specify cssImports.
 *
 * @param plugins - Array of plugin configurations
 * @returns CSS import statements as a string
 *
 * @example
 * ```ts
 * const plugins = [
 *   { packageName: '@test/plugin', cssImports: ['styles.css', 'theme.css'] }
 * ]
 * generateCssImports(plugins)
 * // Returns:
 * // @import 'styles.css';
 * // @import 'theme.css';
 * ```
 */
function generateCssImports(plugins) {
    const cssImports = plugins
        .flatMap((plugin) => plugin.cssImports || [])
        .filter((cssImport) => cssImport && cssImport.trim() !== '')
        .map((cssImport) => `@import '${cssImport}';`);
    return cssImports.join('\n');
}
/**
 * Checks if the generated code is valid (non-empty and has content).
 *
 * @param code - The generated code to check
 * @returns True if code is valid
 */
function isValidCode(code) {
    return typeof code === 'string' && code.trim().length > 0;
}
