// @motiadev/workbench/dist/motia-plugin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = motiaPluginsPlugin;
const core_1 = require("@motiadev/core");
const path_1 = __importDefault(require("path"));
const generator_1 = require("./generator");
const hmr_1 = require("./hmr");
const resolver_1 = require("./resolver");
const types_1 = require("./types");
const utils_1 = require("./utils");
const validator_1 = require("./validator");
/**
 * Vite plugin for loading and managing Motia workbench plugins.
 *
 * Features:
 * - Hot Module Replacement (HMR) support
 * - Runtime validation with detailed error messages
 * - Verbose logging for debugging
 * - CSS injection for plugin styles
 *
 * @param plugins - Array of plugin configurations
 * @param options - Optional loader configuration
 * @returns Vite plugin instance
 *
 * @example
 * ```ts
 * export default defineConfig({
 *   plugins: [
 *     motiaPluginsPlugin([
 *       { packageName: '@my-org/plugin', label: 'My Plugin' }
 *     ])
 *   ]
 * })
 * ```
 */
const printer = new core_1.Printer(process.cwd());
function motiaPluginsPlugin(plugins) {
    let devServer = null;
    try {
        const validationResult = (0, validator_1.validatePlugins)(plugins, {
            failFast: false,
        });
        if (!validationResult.valid) {
            printer.printPluginError('Plugin configuration validation failed:');
            for (const err of validationResult.errors) {
                printer.printPluginError(`  ${err}`);
            }
            throw new Error('Invalid plugin configuration. See errors above.');
        }
        if (validationResult.warnings.length > 0) {
            for (const warning of validationResult.warnings) {
                printer.printPluginWarn(warning);
            }
        }
    }
    catch (error) {
        printer.printPluginError(`Failed to validate plugins: ${error}`);
        throw error;
    }
    const alias = (0, resolver_1.createAliasConfig)(plugins);
    printer.printPluginLog(`Initialized with ${plugins.length} plugin(s)`);
    return {
        name: 'vite-plugin-motia-plugins',
        enforce: 'pre',
        buildStart() {
            printer.printPluginLog('Build started');
        },
        config: () => ({
            resolve: {
                alias,
            },
        }),
        configureServer(server) {
            devServer = server;
            printer.printPluginLog('Dev server configured, HMR enabled');
            const configPaths = [path_1.default.join(process.cwd(), 'motia.config.ts'), path_1.default.join(process.cwd(), 'motia.config.js')];
            for (const configPath of configPaths) {
                server.watcher.add(configPath);
            }
            printer.printPluginLog('Watching for config file changes');
            const localPlugins = plugins.filter((p) => (0, utils_1.isLocalPlugin)(p.packageName));
            if (localPlugins.length > 0) {
                printer.printPluginLog(`Watching ${localPlugins.length} local plugin(s)`);
                for (const plugin of localPlugins) {
                    const resolved = (0, resolver_1.resolvePluginPackage)(plugin);
                    const watchPath = resolved.resolvedPath;
                    server.watcher.add(watchPath);
                    printer.printPluginLog(`Watching: ${watchPath}`);
                }
                server.watcher.on('change', (file) => {
                    const normalizedFile = (0, utils_1.normalizePath)(file);
                    printer.printPluginLog(`File watcher detected change: ${normalizedFile}`);
                });
                server.watcher.on('add', (file) => {
                    const normalizedFile = (0, utils_1.normalizePath)(file);
                    printer.printPluginLog(`File watcher detected new file: ${normalizedFile}`);
                });
            }
        },
        resolveId(id) {
            if (id === types_1.CONSTANTS.VIRTUAL_MODULE_ID) {
                return types_1.CONSTANTS.RESOLVED_VIRTUAL_MODULE_ID;
            }
        },
        load(id) {
            if (id !== types_1.CONSTANTS.RESOLVED_VIRTUAL_MODULE_ID) {
                return null;
            }
            printer.printPluginLog('Loading plugins virtual module');
            printer.printPluginLog('Generating plugin code...');
            const code = (0, generator_1.generatePluginCode)(plugins);
            if (!(0, generator_1.isValidCode)(code)) {
                printer.printPluginError('Generated code is invalid or empty');
                return 'export const plugins = []';
            }
            printer.printPluginLog('Plugin code generated successfully');
            return code;
        },
        async transform(code, id) {
            const normalizedId = (0, utils_1.normalizePath)(id);
            if (!normalizedId.endsWith('src/index.css')) {
                return null;
            }
            printer.printPluginLog('Injecting plugin CSS imports');
            const cssImports = (0, generator_1.generateCssImports)(plugins);
            if (!cssImports) {
                return null;
            }
            return {
                code: `${cssImports}\n${code}`,
                map: null,
            };
        },
        handleHotUpdate(ctx) {
            if (!devServer) {
                printer.printPluginWarn('HMR: Dev server not available');
                return;
            }
            const modulesToUpdate = (0, hmr_1.handlePluginHotUpdate)(ctx, plugins, printer);
            if (modulesToUpdate && modulesToUpdate.length > 0) {
                const merged = Array.from(new Set([...(ctx.modules || []), ...modulesToUpdate]));
                printer.printPluginLog(`HMR: Successfully updated ${merged.length} module(s)`);
                return merged;
            }
        },
        buildEnd() {
            printer.printPluginLog('Build ended');
        },
    };
}
