// @motiadev/workbench/dist/motia-plugin/resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolvePluginPackage = resolvePluginPackage;
exports.createAliasConfig = createAliasConfig;
exports.resolveAllPlugins = resolveAllPlugins;
exports.getUniquePackageNames = getUniquePackageNames;
const utils_1 = require("./utils");
/**
 * Resolves a plugin package to its absolute path and creates an alias.
 *
 * @param plugin - The plugin configuration to resolve
 * @returns Resolved package information including path and alias
 *
 * @example
 * ```ts
 * // Local plugin
 * resolvePluginPackage({ packageName: '~/plugins/my-plugin' })
 * // Returns: {
 * //   packageName: '~/plugins/my-plugin',
 * //   resolvedPath: '/Users/project/plugins/my-plugin',
 * //   isLocal: true,
 * //   alias: '~/plugins/my-plugin'
 * // }
 *
 * // NPM package
 * resolvePluginPackage({ packageName: '@org/plugin' })
 * // Returns: {
 * //   packageName: '@org/plugin',
 * //   resolvedPath: '/Users/project/node_modules/@org/plugin',
 * //   isLocal: false,
 * //   alias: '@org/plugin'
 * // }
 * ```
 */
function resolvePluginPackage(plugin) {
    const { packageName } = plugin;
    const local = (0, utils_1.isLocalPlugin)(packageName);
    const resolvedPath = local ? (0, utils_1.resolveLocalPath)(packageName) : (0, utils_1.resolveNpmPath)(packageName);
    return {
        packageName,
        resolvedPath: (0, utils_1.normalizePath)(resolvedPath),
        isLocal: local,
        alias: packageName,
    };
}
/**
 * Resolves all plugin packages and creates a Vite alias configuration.
 *
 * @param plugins - Array of plugin configurations
 * @returns Vite alias configuration object
 *
 * @example
 * ```ts
 * const plugins = [
 *   { packageName: '~/plugins/local' },
 *   { packageName: '@org/npm-plugin' }
 * ]
 * const aliases = createAliasConfig(plugins)
 * // Returns: {
 * //   '~/plugins/local': '/Users/project/plugins/local',
 * //   '@org/npm-plugin': '/Users/project/node_modules/@org/npm-plugin'
 * // }
 * ```
 */
function createAliasConfig(plugins) {
    // Get unique package names to avoid duplicate aliases
    const uniquePackages = Array.from(new Set(plugins.map((p) => p.packageName)));
    const aliases = {};
    for (const packageName of uniquePackages) {
        const resolved = resolvePluginPackage({ packageName });
        aliases[packageName] = resolved.resolvedPath;
    }
    return aliases;
}
/**
 * Resolves all plugins and returns their resolved package information.
 *
 * @param plugins - Array of plugin configurations
 * @returns Array of resolved package information
 */
function resolveAllPlugins(plugins) {
    return plugins.map((plugin) => resolvePluginPackage(plugin));
}
/**
 * Gets the unique set of package names from plugins.
 *
 * @param plugins - Array of plugin configurations
 * @returns Array of unique package names
 */
function getUniquePackageNames(plugins) {
    return Array.from(new Set(plugins.map((p) => p.packageName)));
}
