// motia/dist/esm/cloud/new-deployment/utils/upload.js
import axios from 'axios';
import fs from 'fs';
import path from 'path';
import { cloudApi } from '../cloud-api';
import { distDir, maxUploadSize } from '../constants';
export const upload = async (deploymentToken, fileName, onProgress) => {
    const filePath = path.join(distDir, fileName);
    const fileStats = fs.statSync(filePath);
    const { fileInfo: { presignedUrl }, } = await cloudApi.upload({
        deploymentToken,
        originalName: fileName,
        size: fileStats.size,
        mimetype: 'application/zip',
    });
    await uploadToPresignedUrl(filePath, presignedUrl, fileStats.size, onProgress);
};
const uploadToPresignedUrl = async (filePath, presignedUrl, fileSize, onProgress) => {
    const fileName = path.basename(filePath);
    const fileStream = fs.createReadStream(filePath);
    await axios.put(presignedUrl, fileStream, {
        headers: {
            'Content-Type': 'application/zip',
            'Content-Length': fileSize,
            'Content-Disposition': `attachment; filename="${fileName}"`,
        },
        maxContentLength: maxUploadSize,
        maxBodyLength: maxUploadSize,
        onUploadProgress(progressEvent) {
            const progress = progressEvent.progress ?? 0;
            onProgress(Math.round(progress * 100));
        },
    });
};
