// @motiadev/workbench/dist/src/components/tutorial/engine/tutorial-engine.js
class Tutorial {
    constructor() {
        Object.defineProperty(this, "steps", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "onStepsRegisteredCallbacks", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "onOpenCallbacks", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
    }
    register(steps) {
        this.steps = steps;
        this.onStepsRegisteredCallbacks.forEach((callback) => callback());
    }
    onStepsRegistered(callback) {
        this.onStepsRegisteredCallbacks.push(callback);
    }
    onOpen(callback) {
        this.onOpenCallbacks.push(callback);
    }
    open() {
        this.onOpenCallbacks.forEach((callback) => callback());
    }
}
export const MotiaTutorial = new Tutorial();
