// @motiadev/workbench/dist/src/components/tutorial/engine/workbench-xpath.js
export const workbenchXPath = {
    sidebarContainer: '//div[@data-testid="sidebar-panel"]',
    closePanelButton: '//div[@id="app-sidebar-container"]//button[@data-testid="close-panel"]',
    bottomPanel: '//div[@id="bottom-panel"]',
    flows: {
        dropdownFlow: (flowId) => `//div[@data-testid="dropdown-${flowId}"]`,
        feature: (featureId) => `//div[@data-feature-id="${featureId}"]`,
        previewButton: (stepId) => `//button[@data-testid="open-code-preview-button-${stepId}"]`,
        node: (stepId) => `//div[@data-testid="node-${stepId}"]`,
    },
    endpoints: {
        endpointsList: '//div[@data-testid="endpoints-list"]',
        endpoint: (method, path) => `//div[@data-testid="endpoint-${method}-${path}"]`,
        callPanel: '//div[@data-testid="endpoint-details-panel"]',
        specButton: '//button[@data-testid="endpoint-spec-button"]',
        bodyTab: '//button[@data-testid="endpoint-body-tab"]',
        headersTab: '//button[@data-testid="endpoint-headers-tab"]',
        paramsTab: '//button[@data-testid="endpoint-params-tab"]',
        callTab: '//button[@data-testid="endpoint-body-tab"]', // deprecated
        response: '//div[@data-testid="endpoint-response-container"]',
        playButton: '//button[@data-testid="endpoint-play-button"]',
    },
    tracing: {
        trace: (index) => `(//button[contains(@class, 'motia-trace-group')])[${index}]`,
        details: '//div[@data-testid="trace-details"]',
        timeline: (index) => `(//div[@data-testid="trace-timeline-item"])[${index}]`,
    },
    logs: {
        container: '//div[@data-testid="logs-container"]',
        searchContainer: '//div[@data-testid="logs-search-container"]',
        traceColumn: (index) => `(//td[starts-with(@data-testid, 'trace')])[${index}]`,
        row: '//div[@data-testid="log-row"]',
    },
    states: {
        container: '//div[@data-testid="states-container"]',
        row: (index) => `(//tr[starts-with(@data-testid, 'item-')])[${index}]`,
    },
    links: {
        flows: '//div[@data-testid="flows-dropdown-trigger"]',
        endpoints: '//button[@data-testid="endpoints-link"]',
        tracing: '//button[@data-testid="tracing-link"]',
        logs: '//button[@data-testid="logs-link"]',
        states: '//button[@data-testid="states-link"]',
    },
};
