// @motiadev/core/dist/src/endpoints/flows-endpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flowsEndpoint = void 0;
const zod_1 = require("zod");
const flows_helper_1 = require("../helper/flows-helper");
const flows_stream_1 = require("../streams/flows-stream");
const flowsEndpoint = (lockedData) => {
    const flowsStream = lockedData.createStream({
        filePath: '__motia.flows',
        hidden: true,
        config: {
            name: '__motia.flows',
            schema: zod_1.z.object({ id: zod_1.z.string(), name: zod_1.z.string(), steps: zod_1.z.any(), edges: zod_1.z.any() }),
            baseConfig: { storageType: 'custom', factory: () => new flows_stream_1.FlowsStream(lockedData) },
        },
    }, { disableTypeCreation: true })();
    lockedData.on('flow-created', (flowId) => {
        const flow = lockedData.flows[flowId];
        const response = (0, flows_helper_1.generateFlow)(flowId, flow.steps);
        flowsStream.set('default', flowId, response);
    });
    lockedData.on('flow-updated', (flowId) => {
        const flow = lockedData.flows[flowId];
        const response = (0, flows_helper_1.generateFlow)(flowId, flow.steps);
        flowsStream.set('default', flowId, response);
    });
    lockedData.on('flow-removed', (flowId) => flowsStream.delete('default', flowId));
};
exports.flowsEndpoint = flowsEndpoint;
