// @motiadev/core/dist/src/endpoints/flows-config-endpoint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.flowsConfigEndpoint = void 0;
const path_1 = __importDefault(require("path"));
const zod_1 = require("zod");
const flows_config_stream_1 = require("../streams/flows-config-stream");
const flowsConfigEndpoint = (app, baseDir, lockedData) => {
    const configPath = path_1.default.join(baseDir, 'motia-workbench.json');
    const stream = new flows_config_stream_1.FlowsConfigStream(configPath);
    lockedData.createStream({
        filePath: '__motia.flowsConfig',
        hidden: true,
        config: {
            name: '__motia.flowsConfig',
            schema: zod_1.z.object({ name: zod_1.z.string(), steps: zod_1.z.any(), edges: zod_1.z.any() }),
            baseConfig: { storageType: 'custom', factory: () => stream },
        },
    }, { disableTypeCreation: true })();
    app.post('/__motia/flows/:id/config', async (req, res) => {
        const newFlowConfig = req.body;
        try {
            await stream.set('default', newFlowConfig.id, newFlowConfig);
            res.status(200).send({ message: 'Flow config saved successfully' });
        }
        catch (error) {
            console.error('Error saving flow config:', error.message);
            res.status(500).json({ error: 'Failed to save flow config' });
        }
    });
};
exports.flowsConfigEndpoint = flowsConfigEndpoint;
