// @motiadev/core/dist/src/streams/flows-config-stream.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowsConfigStream = void 0;
const fs_1 = __importDefault(require("fs"));
const stream_adapter_interface_1 = require("../adapters/interfaces/stream-adapter.interface");
class FlowsConfigStream extends stream_adapter_interface_1.StreamAdapter {
    constructor(configPath) {
        super('__motia.flows-config');
        this.configPath = configPath;
        this.config = [];
    }
    getConfig() {
        if (this.config.length === 0) {
            if (fs_1.default.existsSync(this.configPath)) {
                this.config = JSON.parse(fs_1.default.readFileSync(this.configPath, 'utf8') || '[]');
                if (!Array.isArray(this.config)) {
                    this.config = [];
                    this.setConfig(this.config);
                }
            }
        }
        return this.config;
    }
    setConfig(config) {
        this.config = config;
        fs_1.default.writeFileSync(this.configPath, JSON.stringify(config, null, 2), 'utf8');
    }
    async get(_, id) {
        const allFlowsConfig = this.getConfig();
        return allFlowsConfig.find((flow) => flow.id === id) || null;
    }
    async delete(_) {
        return null;
    }
    async set(_, id, data) {
        const existingConfig = this.getConfig();
        const updatedConfig = existingConfig.filter((item) => item.id !== id);
        updatedConfig.push(data);
        this.setConfig(updatedConfig);
        return data;
    }
    async getGroup() {
        return this.getConfig();
    }
}
exports.FlowsConfigStream = FlowsConfigStream;
