// @motiadev/core/dist/src/streams/redis-logs-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisLogsStream = void 0;
const uuid_1 = require("uuid");
const stream_adapter_interface_1 = require("../adapters/interfaces/stream-adapter.interface");
const LOG_TTL_SECONDS = 3 * 24 * 60 * 60;
class RedisLogsStream extends stream_adapter_interface_1.StreamAdapter {
    constructor(client) {
        super('__motia.logs');
        this.keyPrefix = 'motia:logs:';
        this.client = client;
    }
    makeLogKey(groupId, id) {
        return `${this.keyPrefix}${groupId}:${id}`;
    }
    makeChannelKey(channel) {
        return `${this.keyPrefix}events:${channel.groupId}:${channel.id}`;
    }
    makeLogId() {
        return `${Date.now()}-${(0, uuid_1.v4)()}`;
    }
    async scanKeys(pattern) {
        const keys = [];
        let cursor = '0';
        do {
            const result = await this.client.scan(cursor.toString(), {
                MATCH: pattern,
                COUNT: 100,
            });
            cursor = result.cursor;
            keys.push(...result.keys);
        } while (String(cursor) !== '0');
        return keys;
    }
    async get(groupId, id) {
        const key = this.makeLogKey(groupId, id);
        const value = await this.client.get(key);
        return value ? JSON.parse(value) : null;
    }
    async set(groupId, id, data) {
        const logId = id || this.makeLogId();
        const key = this.makeLogKey(groupId, logId);
        const logData = { ...data, id: logId };
        const item = { ...logData, id: logId };
        const itemJson = JSON.stringify(item);
        await Promise.all([
            this.client.set(key, itemJson),
            this.client.expire(key, LOG_TTL_SECONDS),
            this.send({ groupId, id: logId }, { type: 'log', data: item }),
        ]);
        return item;
    }
    async delete(groupId, id) {
        const key = this.makeLogKey(groupId, id);
        const value = await this.client.get(key);
        if (!value)
            return null;
        const item = JSON.parse(value);
        await Promise.all([this.client.del(key), this.send({ groupId, id }, { type: 'delete', data: item })]);
        return item;
    }
    async getGroup(groupId) {
        const pattern = `${this.keyPrefix}${groupId}:*`;
        const keys = await this.scanKeys(pattern);
        if (keys.length === 0)
            return [];
        const values = await this.client.mGet(keys);
        const logs = values.filter((v) => v !== null).map((v) => JSON.parse(v));
        const sortDesc = (a, b) => a.time - b.time;
        return logs.sort(sortDesc);
    }
    async send(channel, event) {
        const channelKey = this.makeChannelKey(channel);
        await this.client.publish(channelKey, JSON.stringify(event));
    }
    async clear(groupId) {
        const pattern = `${this.keyPrefix}${groupId}:*`;
        const keys = await this.scanKeys(pattern);
        if (keys.length === 0)
            return;
        await this.client.del(keys);
    }
}
exports.RedisLogsStream = RedisLogsStream;
