// motia/dist/cjs/cloud/new-deployment/utils/upload.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.upload = void 0;
const axios_1 = __importDefault(require("axios"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const cloud_api_1 = require("../cloud-api");
const constants_1 = require("../constants");
const upload = async (deploymentToken, fileName, onProgress) => {
    const filePath = path_1.default.join(constants_1.distDir, fileName);
    const fileStats = fs_1.default.statSync(filePath);
    const { fileInfo: { presignedUrl }, } = await cloud_api_1.cloudApi.upload({
        deploymentToken,
        originalName: fileName,
        size: fileStats.size,
        mimetype: 'application/zip',
    });
    await uploadToPresignedUrl(filePath, presignedUrl, fileStats.size, onProgress);
};
exports.upload = upload;
const uploadToPresignedUrl = async (filePath, presignedUrl, fileSize, onProgress) => {
    const fileName = path_1.default.basename(filePath);
    const fileStream = fs_1.default.createReadStream(filePath);
    await axios_1.default.put(presignedUrl, fileStream, {
        headers: {
            'Content-Type': 'application/zip',
            'Content-Length': fileSize,
            'Content-Disposition': `attachment; filename="${fileName}"`,
        },
        maxContentLength: constants_1.maxUploadSize,
        maxBodyLength: constants_1.maxUploadSize,
        onUploadProgress(progressEvent) {
            const progress = progressEvent.progress ?? 0;
            onProgress(Math.round(progress * 100));
        },
    });
};
