// motia/dist/cjs/cloud/cli-output-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIOutputManager = exports.Message = void 0;
const colors_1 = __importDefault(require("colors"));
const readline_1 = __importDefault(require("readline"));
const table_1 = require("table");
const progress = colors_1.default.yellow('➜ [PROGRESS]');
const success = colors_1.default.green('✓ [SUCCESS]');
const failed = colors_1.default.red('✘ [ERROR]');
const warning = colors_1.default.yellow('! [WARNING]');
const info = colors_1.default.blue('i [INFO]');
const tags = {
    success,
    failed,
    progress,
    warning,
    info,
};
const colorTags = {
    gray: colors_1.default.gray,
    dark: colors_1.default.magenta,
    red: colors_1.default.red,
    green: colors_1.default.green,
    yellow: colors_1.default.yellow,
    blue: colors_1.default.blue,
    magenta: colors_1.default.magenta,
    cyan: colors_1.default.cyan,
    white: colors_1.default.white,
};
class Message {
    constructor() {
        this.output = [];
    }
    dark(message) {
        return colors_1.default.magenta(message);
    }
    text(message) {
        this.output.push(message);
        return this;
    }
    tag(tag) {
        this.output.push(tags[tag]);
        return this;
    }
    append(message, color) {
        if (color) {
            this.output.push(colorTags[color](message));
        }
        else {
            this.output.push(message);
        }
        return this;
    }
    box(messages, color = 'blue') {
        const message = messages.join('\n \n');
        const lines = message.match(/.{1,40}/g) || [message];
        const width = Math.min(40, Math.max(...lines.map((line) => line.length)));
        const border = '─'.repeat(width + 2);
        const borderColor = colorTags[color];
        this.output.push(borderColor('\n ┌' + border + '┐\n'));
        lines.forEach((line) => {
            const padding = ' '.repeat(width - line.trim().length);
            this.output.push(borderColor('│ ') + line.trim() + padding + borderColor(' │\n'));
        });
        this.output.push(borderColor('└' + border + '┘'));
        return this;
    }
    table(headers, rows) {
        const colouredHeaders = headers?.map((header) => colors_1.default.blue(colors_1.default.bold(header)));
        const records = [colouredHeaders, ...rows].filter((record) => record !== undefined);
        this.output.push((0, table_1.table)(records, {
            border: {
                topBody: colors_1.default.blue('─'),
                topJoin: colors_1.default.blue('┬'),
                topLeft: colors_1.default.blue('┌'),
                topRight: colors_1.default.blue('┐'),
                bodyLeft: colors_1.default.blue('│'),
                bodyRight: colors_1.default.blue('│'),
                bottomBody: colors_1.default.blue('─'),
                bottomJoin: colors_1.default.blue('┴'),
                bottomLeft: colors_1.default.blue('└'),
                bottomRight: colors_1.default.blue('┘'),
                joinLeft: colors_1.default.blue('├'),
                joinRight: colors_1.default.blue('┤'),
                joinMiddleDown: colors_1.default.blue('│'),
                joinMiddleUp: colors_1.default.blue(''),
                joinMiddleLeft: colors_1.default.blue('│'),
                joinMiddleRight: colors_1.default.blue('│'),
                bodyJoin: colors_1.default.blue('│'),
                joinBody: colors_1.default.blue('─'),
                headerJoin: colors_1.default.blue('│'),
                joinJoin: colors_1.default.blue('┼'),
            },
        }));
        return this;
    }
    toString() {
        return this.output.join(' ');
    }
}
exports.Message = Message;
class CLIOutputManager {
    constructor() {
        this.lines = new Map(); // Track line positions
        this.lineCount = 0;
    }
    log(id, callback) {
        const lineIndex = this.lines.get(id);
        if (lineIndex === undefined) {
            const msg = this.createMessage();
            callback(msg);
            this.lines.set(id, this.lineCount);
            process.stdout.write(`${msg.toString()}\n`);
            this.lineCount++;
            return;
        }
        const msg = this.createMessage();
        callback(msg);
        readline_1.default.moveCursor(process.stdout, 0, -(this.lineCount - lineIndex));
        readline_1.default.clearLine(process.stdout, 0);
        process.stdout.write(`${msg.toString()}\n`);
        readline_1.default.moveCursor(process.stdout, 0, this.lineCount - lineIndex - 1);
    }
    createMessage() {
        return new Message();
    }
}
exports.CLIOutputManager = CLIOutputManager;
