// motia/dist/cjs/cloud/config-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliContext = void 0;
exports.handler = handler;
const analytics_node_1 = require("@amplitude/analytics-node");
const analytics_1 = require("../utils/analytics");
const cli_output_manager_1 = require("./cli-output-manager");
const getCommandName = () => {
    const args = process.argv.slice(2);
    const commandParts = [];
    for (let i = 0; i < args.length && i < 3; i++) {
        const arg = args[i];
        if (!arg.startsWith('-') && !arg.startsWith('--')) {
            commandParts.push(arg);
        }
        else {
            break;
        }
    }
    return commandParts.join(' ') || 'unknown';
};
class CliContext {
    constructor() {
        this.output = new cli_output_manager_1.CLIOutputManager();
    }
    log(id, callback) {
        this.output.log(id, callback);
    }
    exitWithError(msg, error) {
        this.output.log('error', (message) => {
            message.tag('failed').append(msg);
            if (error) {
                message.box([error instanceof Error ? error.message : 'Unknown error'], 'red');
            }
        });
        process.exit(1);
    }
    exit(code) {
        process.exit(code);
    }
}
exports.CliContext = CliContext;
function handler(handler) {
    return async (args) => {
        const context = new CliContext();
        const commandName = getCommandName();
        try {
            await handler(args, context);
        }
        catch (error) {
            (0, analytics_1.logCliError)(commandName, error);
            await (0, analytics_node_1.flush)().promise.catch(() => {
                // Silently fail
            });
            if (error instanceof Error) {
                context.log('error', (message) => message.tag('failed').append(error.message));
                context.exit(1);
            }
            else {
                context.exitWithError('An error occurred', error);
            }
        }
    };
}
