// motia/dist/cjs/cloud/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployEndpoints = void 0;
const crypto_1 = require("crypto");
const build_validation_1 = require("./build/build-validation");
const build_1 = require("./new-deployment/build");
const cloud_api_1 = require("./new-deployment/cloud-api");
const streaming_deployment_listener_1 = require("./new-deployment/listeners/streaming-deployment-listener");
const deployment_stream_1 = require("./new-deployment/streams/deployment-stream");
const upload_artifacts_1 = require("./new-deployment/upload-artifacts");
const deployEndpoints = (server, lockedData) => {
    const { app } = server;
    // Criar stream de deployment se não existir
    const deploymentStream = lockedData.createStream({
        filePath: '__motia.deployment',
        hidden: true,
        config: {
            name: '__motia.deployment',
            baseConfig: { storageType: 'default' },
            schema: null,
        },
    })();
    const deploymentManager = new deployment_stream_1.DeploymentStreamManager(deploymentStream);
    app.post('/__motia/cloud/deploy/start', async (req, res) => {
        try {
            const { deploymentToken, deploymentId, envs } = req.body;
            const sessionId = deploymentId || (0, crypto_1.randomUUID)();
            if (!deploymentToken || !deploymentId) {
                return res.status(400).json({
                    success: false,
                    error: 'deploymentToken and deploymentId are required',
                });
            }
            await deploymentManager.startDeployment(sessionId);
            const listener = new streaming_deployment_listener_1.StreamingDeploymentListener(sessionId, deploymentStream);
            res.json({
                success: true,
                message: 'Deployment started',
                deploymentId: sessionId,
                streamName: 'deployment-status',
                groupId: 'deployments',
                itemId: sessionId,
            });
            setImmediate(async () => {
                try {
                    await listener.startBuildPhase();
                    const builder = await (0, build_1.build)(listener).catch(() => {
                        throw new Error('Build failed, check the logs for more information');
                    });
                    const isValid = (0, build_validation_1.buildValidation)(builder, listener);
                    if (!isValid) {
                        await listener.onBuildErrors(listener.getErrors());
                        return;
                    }
                    await listener.startUploadPhase();
                    await (0, upload_artifacts_1.uploadArtifacts)(builder, deploymentToken, listener);
                    await listener.startDeployPhase();
                    await cloud_api_1.cloudApi.startDeployment({
                        deploymentToken,
                        envVars: envs,
                        steps: builder.stepsConfig,
                        streams: builder.streamsConfig,
                        routers: builder.routersConfig,
                    });
                }
                catch (error) {
                    console.error('Deployment failed:', error);
                    // Update stream with error
                    if (listener) {
                        await listener.onDeployError(error.message);
                    }
                }
            });
        }
        catch (error) {
            console.error('Failed to start deployment:', error);
            res.status(500).json({ success: false, error: error.message });
        }
    });
    app.get('/__motia/cloud/deploy/status/:deploymentId', async (req, res) => {
        try {
            const { deploymentId } = req.params;
            const deployment = await deploymentManager.getDeployment(deploymentId);
            return deployment
                ? res.status(200).json({ success: true, deployment })
                : res.status(404).json({ success: false, error: 'Deployment not found' });
        }
        catch (error) {
            res.status(500).json({ success: false, error: error.message });
        }
    });
};
exports.deployEndpoints = deployEndpoints;
