// motia/dist/cjs/openapi/generate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateOpenApi = generateOpenApi;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const process_schema_1 = require("./process-schema");
const utils_1 = require("./utils");
function generateOpenApi(projectDir, apiSteps, title, version, outputFile = 'openapi.json') {
    // read package.json to get project name for default title & version
    if (!title || !version) {
        try {
            const packageJsonPath = path.join(projectDir, 'package.json');
            const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
            if (!title)
                title = packageJson.name;
            if (!version)
                version = packageJson.version;
        }
        catch (error) {
            console.warn(`Could not read package.json in ${projectDir} to determine project name. Using default.`);
        }
    }
    const openApi = {
        openapi: '3.0.0',
        info: {
            title: title ?? 'Motia Project API',
            version: version ?? '1.0.0',
        },
        paths: {},
        components: {
            schemas: {},
        },
    };
    for (const step of apiSteps) {
        const pathItem = openApi.paths[step.config.path] || {};
        const method = step.config.method.toLowerCase();
        const operation = {
            summary: step.config.name,
            description: step.config.description,
            requestBody: undefined,
            responses: {},
        };
        if (step.config.queryParams) {
            operation.parameters = operation.parameters || [];
            for (const param of step.config.queryParams) {
                operation.parameters.push({
                    in: 'query',
                    name: param.name,
                    description: param.description,
                    required: false,
                    schema: {
                        type: 'string',
                    },
                });
            }
        }
        if (step.config.bodySchema) {
            const bodySchema = step.config.bodySchema;
            delete bodySchema.$schema;
            const processedSchema = (0, process_schema_1.processSchema)(bodySchema, openApi);
            operation.requestBody = {
                content: {
                    'application/json': {
                        schema: processedSchema,
                    },
                },
            };
        }
        if (step.config.responseSchema) {
            for (const [statusCode, responseSchema] of Object.entries(step.config.responseSchema)) {
                const resSchema = responseSchema;
                delete resSchema.$schema;
                const processedSchema = (0, process_schema_1.processSchema)(resSchema, openApi);
                operation.responses[statusCode] = {
                    description: `Response for status code ${statusCode}`,
                    content: {
                        'application/json': {
                            schema: processedSchema,
                        },
                    },
                };
            }
        }
        if ((0, utils_1.isHttpMethod)(method)) {
            pathItem[method] = operation;
        }
        openApi.paths[step.config.path] = pathItem;
    }
    const openApiJson = JSON.stringify(openApi, null, 2);
    fs.writeFileSync(path.join(projectDir, outputFile), openApiJson);
    console.log(`✅ OpenAPI specification generated successfully at ${outputFile}`);
}
