// @motiadev/core/dist/src/infrastructure-validator/validations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateInfrastructureConfig = exports.validateQueueConfig = void 0;
const zod_1 = require("zod");
const schemas_1 = require("./schemas");
const validateQueueConfig = (queueConfig) => {
    try {
        const schema = schemas_1.queueSchema;
        schema.parse(queueConfig);
        return { success: true };
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            const errors = error.issues.map((err) => ({
                path: err.path.length > 0 ? err.path.join('.') : 'queue',
                message: err.message,
            }));
            return {
                success: false,
                errors,
            };
        }
        return {
            success: false,
            errors: [
                {
                    path: 'queue',
                    message: `Unexpected validation error: ${error instanceof Error ? error.message : 'Unknown error'}`,
                },
            ],
        };
    }
};
exports.validateQueueConfig = validateQueueConfig;
const validateInfrastructureConfig = (infrastructureConfig) => {
    try {
        schemas_1.infrastructureSchema.parse(infrastructureConfig);
        return { success: true };
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            const errors = error.issues.map((err) => ({
                path: err.path.length > 0 ? err.path.join('.') : 'infrastructure',
                message: err.message,
            }));
            return {
                success: false,
                errors,
            };
        }
        return {
            success: false,
            errors: [
                {
                    path: 'infrastructure',
                    message: `Unexpected validation error: ${error instanceof Error ? error.message : 'Unknown error'}`,
                },
            ],
        };
    }
};
exports.validateInfrastructureConfig = validateInfrastructureConfig;
