// motia/dist/esm/cloud/config-utils.js
import { flush } from '@amplitude/analytics-node';
import { logCliError } from '../utils/analytics';
import { CLIOutputManager } from './cli-output-manager';
const getCommandName = () => {
    const args = process.argv.slice(2);
    const commandParts = [];
    for (let i = 0; i < args.length && i < 3; i++) {
        const arg = args[i];
        if (!arg.startsWith('-') && !arg.startsWith('--')) {
            commandParts.push(arg);
        }
        else {
            break;
        }
    }
    return commandParts.join(' ') || 'unknown';
};
export class CliContext {
    constructor() {
        this.output = new CLIOutputManager();
    }
    log(id, callback) {
        this.output.log(id, callback);
    }
    exitWithError(msg, error) {
        this.output.log('error', (message) => {
            message.tag('failed').append(msg);
            if (error) {
                message.box([error instanceof Error ? error.message : 'Unknown error'], 'red');
            }
        });
        process.exit(1);
    }
    exit(code) {
        process.exit(code);
    }
}
export function handler(handler) {
    return async (args) => {
        const context = new CliContext();
        const commandName = getCommandName();
        try {
            await handler(args, context);
        }
        catch (error) {
            logCliError(commandName, error);
            await flush().promise.catch(() => {
                // Silently fail
            });
            if (error instanceof Error) {
                context.log('error', (message) => message.tag('failed').append(error.message));
                context.exit(1);
            }
            else {
                context.exitWithError('An error occurred', error);
            }
        }
    };
}
