// @motiadev/core/dist/src/analytics/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackEvent = exports.isAnalyticsEnabled = exports.getProjectIdentifier = exports.getUserIdentifier = exports.getProjectName = void 0;
const analytics_node_1 = require("@amplitude/analytics-node");
const crypto_1 = __importDefault(require("crypto"));
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const getProjectName = (baseDir) => {
    const packageJsonPath = path_1.default.join(baseDir, 'package.json');
    if (fs_1.default.existsSync(packageJsonPath)) {
        const packageJson = JSON.parse(fs_1.default.readFileSync(packageJsonPath, 'utf8'));
        return packageJson.name || path_1.default.basename(baseDir);
    }
    return 'unknown';
};
exports.getProjectName = getProjectName;
const getUserIdentifier = () => {
    const userInfo = `${os_1.default.userInfo().username}${os_1.default.hostname()}`;
    return crypto_1.default.createHash('sha256').update(userInfo).digest('hex').substring(0, 16);
};
exports.getUserIdentifier = getUserIdentifier;
const getProjectIdentifier = (baseDir) => {
    try {
        return crypto_1.default.createHash('sha256').update((0, exports.getProjectName)(baseDir)).digest('hex').substring(0, 16);
    }
    catch (error) {
        return 'unknown';
    }
};
exports.getProjectIdentifier = getProjectIdentifier;
const isAnalyticsEnabled = () => {
    return process.env.MOTIA_ANALYTICS_DISABLED !== 'true';
};
exports.isAnalyticsEnabled = isAnalyticsEnabled;
const trackEvent = (eventName, properties = {}) => {
    try {
        if ((0, exports.isAnalyticsEnabled)()) {
            (0, analytics_node_1.track)(eventName, properties, {
                user_id: (0, exports.getUserIdentifier)() || 'unknown',
            });
        }
    }
    catch (error) {
        // Silently fail to not disrupt dev server
    }
};
exports.trackEvent = trackEvent;
