// @motiadev/core/dist/src/observability/tracer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTracerFactory = exports.BaseTracerFactory = void 0;
const create_trace_1 = require("./create-trace");
const redis_trace_stream_adapter_1 = require("./redis-trace-stream-adapter");
const stream_tracer_1 = require("./stream-tracer");
const trace_manager_1 = require("./trace-manager");
const MAX_TRACE_GROUPS = process.env.MOTIA_MAX_TRACE_GROUPS //
    ? Number.parseInt(process.env.MOTIA_MAX_TRACE_GROUPS, 10)
    : 50;
class BaseTracerFactory {
    constructor(traceStream, traceGroupStream) {
        this.traceStream = traceStream;
        this.traceGroupStream = traceGroupStream;
    }
    async getAllGroups() {
        return await this.traceGroupStream.getGroup('default');
    }
    async deleteGroup(group) {
        const traces = await this.traceStream.getGroup(group.id);
        for (const trace of traces) {
            await this.traceStream.delete(group.id, trace.id);
        }
        await this.traceGroupStream.delete('default', group.id);
    }
    async clear() {
        const groups = await this.getAllGroups();
        for (const group of groups) {
            await this.deleteGroup(group);
        }
    }
    async createTracer(traceId, step, logger) {
        const traceGroup = {
            id: traceId,
            name: step.config.name,
            lastActivity: Date.now(),
            metadata: {
                completedSteps: 0,
                activeSteps: 0,
                totalSteps: 0,
            },
            correlationId: undefined,
            status: 'running',
            startTime: Date.now(),
        };
        const groups = await this.getAllGroups();
        if (groups.length >= MAX_TRACE_GROUPS) {
            const groupsToDelete = groups
                .sort((a, b) => a.startTime - b.startTime) // date ascending
                .slice(0, groups.length - MAX_TRACE_GROUPS + 1);
            for (const group of groupsToDelete) {
                await this.deleteGroup(group);
            }
        }
        const trace = (0, create_trace_1.createTrace)(traceGroup, step);
        const manager = new trace_manager_1.TraceManager(this.traceStream, this.traceGroupStream, traceGroup, trace);
        return new stream_tracer_1.StreamTracer(manager, traceGroup, trace, logger);
    }
    async attachToTrace(traceId, step, logger) {
        const existingGroup = await this.traceGroupStream.get('default', traceId);
        if (!existingGroup) {
            return this.createTracer(traceId, step, logger);
        }
        const trace = (0, create_trace_1.createTrace)(existingGroup, step);
        const manager = new trace_manager_1.TraceManager(this.traceStream, this.traceGroupStream, existingGroup, trace);
        return new stream_tracer_1.StreamTracer(manager, existingGroup, trace, logger);
    }
}
exports.BaseTracerFactory = BaseTracerFactory;
const createTracerFactory = (lockedData) => {
    const traceStreamName = 'motia-trace';
    const traceStreamAdapter = new redis_trace_stream_adapter_1.RedisTraceStreamAdapter(traceStreamName, lockedData.redisClient);
    const traceStream = lockedData.createStream({
        filePath: traceStreamName,
        hidden: true,
        config: {
            name: traceStreamName,
            baseConfig: { storageType: 'custom', factory: () => traceStreamAdapter },
            schema: null,
        },
    })();
    const traceGroupName = 'motia-trace-group';
    const traceGroupStreamAdapter = new redis_trace_stream_adapter_1.RedisTraceStreamAdapter(traceGroupName, lockedData.redisClient);
    const traceGroupStream = lockedData.createStream({
        filePath: traceGroupName,
        hidden: true,
        config: {
            name: traceGroupName,
            baseConfig: { storageType: 'custom', factory: () => traceGroupStreamAdapter },
            schema: null,
        },
    })();
    return new BaseTracerFactory(traceStream, traceGroupStream);
};
exports.createTracerFactory = createTracerFactory;
