// motia/dist/cjs/cloud/build/builders/node/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeBuilder = void 0;
const colors_1 = __importDefault(require("colors"));
const esbuild = __importStar(require("esbuild"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const constants_1 = require("../../../new-deployment/constants");
const archiver_1 = require("../archiver");
const include_static_files_1 = require("../include-static-files");
class NodeBuilder {
    constructor(builder, listener) {
        this.builder = builder;
        this.listener = listener;
    }
    async loadEsbuildConfig() {
        const configFiles = ['esbuild.config.json', '.esbuildrc.json'];
        for (const configFile of configFiles) {
            const configPath = path_1.default.join(this.builder.projectDir, configFile);
            if (fs_1.default.existsSync(configPath)) {
                try {
                    const configContent = fs_1.default.readFileSync(configPath, 'utf-8');
                    return JSON.parse(configContent);
                }
                catch (err) {
                    console.warn(colors_1.default.yellow(`Warning: Failed to load esbuild config from ${configFile}`));
                }
            }
        }
        return null;
    }
    async buildApiSteps(steps) {
        const relativePath = path_1.default.relative(constants_1.distDir, this.builder.projectDir);
        const getStepPath = (step) => {
            return step.filePath
                .replace(this.builder.projectDir, relativePath)
                .replace(/(.*)\.(ts|js)$/, '$1.js')
                .replace(/\\/g, '/');
        };
        const file = fs_1.default
            .readFileSync(path_1.default.join(__dirname, 'router-template.ts'), 'utf-8')
            .replace('// {{imports}}', steps.map((step, index) => `import * as route${index} from '${getStepPath(step)}'`).join('\n'))
            .replace('// {{router paths}}', steps
            .map((step, index) => `'${step.config.method} ${step.config.path}': { stepName: '${step.config.name}', handler: route${index}.handler, config: route${index}.config }`)
            .join(',\n'));
        const tsRouter = path_1.default.join(constants_1.distDir, 'router.ts');
        fs_1.default.writeFileSync(tsRouter, file);
        const userConfig = await this.loadEsbuildConfig();
        const defaultConfig = {
            entryPoints: [tsRouter],
            bundle: true,
            sourcemap: true,
            outfile: path_1.default.join(constants_1.distDir, 'router.js'),
            platform: 'node',
        };
        await esbuild.build(userConfig ? { ...defaultConfig, ...userConfig } : defaultConfig);
        const zipName = 'router-node.zip';
        const archiver = new archiver_1.Archiver(path_1.default.join(constants_1.distDir, zipName));
        const routerJs = path_1.default.join(constants_1.distDir, 'router.js');
        const routerMap = path_1.default.join(constants_1.distDir, 'router.js.map');
        archiver.append(fs_1.default.readFileSync(routerJs), 'router.js');
        archiver.append(fs_1.default.readFileSync(routerMap), 'router.js.map');
        (0, include_static_files_1.includeStaticFiles)(steps, this.builder, archiver);
        const { compressedSize, uncompressedSize } = await archiver.finalize();
        fs_1.default.unlinkSync(tsRouter);
        fs_1.default.unlinkSync(routerJs);
        fs_1.default.unlinkSync(routerMap);
        return { compressedSize, uncompressedSize, path: zipName };
    }
    async build(step) {
        const relativeFilePath = step.filePath.replace(this.builder.projectDir, '');
        const entrypointPath = relativeFilePath.replace(/(.*)\.(ts|js)$/, '$1.js');
        const entrypointMapPath = entrypointPath.replace(/(.*)\.js$/, '$1.js.map');
        const bundlePath = path_1.default.join('node', entrypointPath.replace(/(.*)\.js$/, '$1.zip'));
        const outputJsFile = path_1.default.join(constants_1.distDir, 'node', entrypointPath);
        const outputMapFile = path_1.default.join(constants_1.distDir, 'node', entrypointMapPath);
        this.builder.registerStep({ entrypointPath, bundlePath, step, type: 'node' });
        this.listener.onBuildStart(step);
        try {
            const userConfig = await this.loadEsbuildConfig();
            const defaultConfig = {
                entryPoints: [step.filePath],
                bundle: true,
                sourcemap: true,
                outfile: outputJsFile,
                platform: 'node',
            };
            await esbuild.build(userConfig ? { ...defaultConfig, ...userConfig } : defaultConfig);
            const archiver = new archiver_1.Archiver(path_1.default.join(constants_1.distDir, bundlePath));
            archiver.append(fs_1.default.readFileSync(outputJsFile), entrypointPath);
            archiver.append(fs_1.default.readFileSync(outputMapFile), entrypointMapPath);
            (0, include_static_files_1.includeStaticFiles)([step], this.builder, archiver);
            const { compressedSize, uncompressedSize } = await archiver.finalize();
            fs_1.default.unlinkSync(outputJsFile);
            fs_1.default.unlinkSync(outputMapFile);
            this.builder.recordStepSize(step, compressedSize, uncompressedSize);
            this.listener.onBuildEnd(step, compressedSize);
        }
        catch (err) {
            this.listener.onBuildError(step, err);
            throw err;
        }
    }
}
exports.NodeBuilder = NodeBuilder;
