// @motiadev/stream-client-browser/dist/src/stream-adapter.js
export class StreamSocketAdapter {
    constructor(address) {
        this.address = address;
        this.onMessageListeners = [];
        this.onOpenListeners = [];
        this.onCloseListeners = [];
        this.ws = new WebSocket(this.address);
    }
    connect() { }
    send(message) {
        this.ws.send(message);
    }
    onMessage(callback) {
        const listener = (message) => callback(message.data);
        this.ws.addEventListener('message', listener);
        this.onMessageListeners.push(listener);
    }
    onOpen(callback) {
        this.ws.addEventListener('open', callback);
        this.onOpenListeners.push(callback);
    }
    onClose(callback) {
        this.ws.addEventListener('close', callback);
        this.onCloseListeners.push(callback);
    }
    close() {
        this.ws.close();
        this.onMessageListeners.forEach((listener) => this.ws.removeEventListener('message', listener));
        this.onOpenListeners.forEach((listener) => this.ws.removeEventListener('open', listener));
        this.onCloseListeners.forEach((listener) => this.ws.removeEventListener('close', listener));
    }
    isOpen() {
        return this.ws.readyState === WebSocket.OPEN;
    }
}
