// @motiadev/stream-client/dist/cjs/__tests__/stream.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const stream_1 = require("../src/stream");
class MockSocket {
    constructor() {
        this.listeners = {};
    }
    addEventListener(event, cb) {
        if (!this.listeners[event])
            this.listeners[event] = [];
        this.listeners[event].push(cb);
    }
    connect() { }
    isOpen() {
        return true;
    }
    onMessage(callback) {
        this.addEventListener('message', callback);
    }
    onOpen(callback) {
        this.addEventListener('open', callback);
    }
    onClose(callback) {
        this.addEventListener('close', callback);
    }
    send(data) {
        this.listeners['message']?.forEach((cb) => cb(data));
    }
    close() {
        this.listeners = {};
    }
}
function getSocket(stream) {
    return stream.ws;
}
function makeGroupMessage(type, data, timestamp = Date.now()) {
    return {
        streamName: 'test-stream',
        groupId: 'test-group',
        timestamp,
        event: { type, ...(type === 'event' ? { event: data } : { data }) },
    };
}
function makeItemMessage(type, data, timestamp = Date.now()) {
    return {
        streamName: 'test-stream',
        groupId: 'test-group',
        id: '1',
        timestamp,
        event: { type, ...(type === 'event' ? { event: data } : { data }) },
    };
}
describe('Stream', () => {
    beforeEach(() => {
        global.WebSocket = MockSocket;
    });
    afterEach(() => {
        delete global.WebSocket;
    });
    it('should sync group events', () => {
        const stream = new stream_1.Stream(() => new MockSocket());
        const socket = getSocket(stream);
        const sub = stream.subscribeGroup('test-stream', 'test-group');
        const syncData = [
            { id: '1', name: 'A' },
            { id: '2', name: 'B' },
        ];
        socket.send(JSON.stringify(makeGroupMessage('sync', syncData)));
        expect(sub.getState()).toEqual(syncData);
    });
    it('should not sync item events in group subscriptions', () => {
        const stream = new stream_1.Stream(() => new MockSocket());
        const socket = getSocket(stream);
        const sub = stream.subscribeGroup('test-stream', 'test-group');
        const syncData = [
            { id: '1', name: 'A' },
            { id: '2', name: 'B' },
        ];
        socket.send(JSON.stringify(makeGroupMessage('sync', syncData)));
        socket.send(JSON.stringify(makeItemMessage('sync', syncData[0])));
        expect(sub.getState()).toEqual(syncData);
    });
});
