// motia/dist/esm/cloud/new-deployment/listeners/build-printer.js
import { Printer } from '@motiadev/core/dist/src/printer';
import colors from 'colors';
import { CLIOutputManager } from '../../cli-output-manager';
import { prettyBytes } from '../utils/pretty-bytes';
const building = colors.yellow('➜ [BUILDING]');
const built = colors.green('✓ [BUILT]');
const failed = colors.red('✘ [FAILED]');
const skipped = colors.gray('- [SKIPPED]');
const baseTag = (tag) => colors.bold(colors.magenta(tag));
const streamTag = baseTag('Stream');
const stepTag = baseTag('Step');
const routerTag = baseTag('Router');
export class BuildPrinter {
    constructor() {
        this.printer = new Printer(process.cwd());
        this.output = new CLIOutputManager();
    }
    getLanguage(language) {
        if (language === 'python') {
            return colors.bold(colors.blue('Python'));
        }
        else if (language === 'node') {
            return colors.bold(colors.green('Node'));
        }
        else if (language === 'ruby') {
            return colors.bold(colors.red('Ruby'));
        }
        return colors.bold(colors.gray('Unknown'));
    }
    getStepLanguage(step) {
        if (step.filePath.endsWith('.py')) {
            return this.getLanguage('python');
        }
        else if (step.filePath.endsWith('.js') || step.filePath.endsWith('.ts')) {
            return this.getLanguage('node');
        }
        else if (step.filePath.endsWith('.rb')) {
            return this.getLanguage('ruby');
        }
        return this.getLanguage('unknown');
    }
    printStepBuilding(step, progressMessage) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => {
            message.append(`${building} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`);
            if (progressMessage) {
                message.append(colors.yellow(progressMessage));
            }
        });
    }
    printStepBuilt(step, size) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message
            .append(`${built} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`)
            .append(`${prettyBytes(size)}`, 'gray'));
    }
    printApiRouterBuilding(language) {
        const fileName = `router-${language}.zip`;
        const coloredFileName = colors.bold(colors.cyan(fileName));
        this.output.log(fileName, (message) => message //
            .append(routerTag)
            .append(building)
            .append(this.getLanguage(language))
            .append(coloredFileName));
    }
    printApiRouterBuilt(language, size) {
        const fileName = `router-${language}.zip`;
        const coloredFileName = colors.bold(colors.cyan(fileName));
        this.output.log(fileName, (message) => message
            .append(built)
            .append(routerTag)
            .append(this.getLanguage(language))
            .append(coloredFileName)
            .append(prettyBytes(size), 'gray'));
    }
    printStepFailed(step, error) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message.append(`${failed} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`).append(error.message, 'red'));
    }
    printStepSkipped(step, reason) {
        const stepLanguage = this.getStepLanguage(step);
        const stepType = this.printer.getStepType(step);
        const stepPath = this.printer.getStepPath(step);
        this.output.log(step.filePath, (message) => message.append(`${skipped} ${stepTag} ${stepLanguage} ${stepType} ${stepPath}`).append(reason, 'yellow'));
    }
    printStreamCreated(stream) {
        const streamPath = this.getStreamPath(stream);
        const streamName = colors.gray(`[${stream.config.name}]`);
        this.output.log(stream.filePath, (message) => message.append(`${built} ${streamTag} ${streamName} ${streamPath}`));
    }
    getStreamPath(stream) {
        return colors.bold(colors.cyan(this.printer.getRelativePath(stream.filePath)));
    }
}
