// motia/dist/esm/cloud/new-deployment/listeners/printer.js
import { Printer } from '@motiadev/core/dist/src/printer';
import colors from 'colors';
import { CLIOutputManager } from '../../cli-output-manager';
import { projectDir } from '../constants';
const uploading = colors.yellow('➜ [UPLOADING]');
const uploaded = colors.green('✓ [UPLOADED]');
export class DeployPrinter {
    constructor() {
        this.printer = new Printer(projectDir);
        this.output = new CLIOutputManager();
    }
    getLanguage(language) {
        if (language === 'node') {
            return colors.bold(colors.green('Node'));
        }
        else if (language === 'python') {
            return colors.bold(colors.blue('Python'));
        }
        else if (language === 'ruby') {
            return colors.bold(colors.red('Ruby'));
        }
        return colors.bold(colors.gray('Unknown'));
    }
    getStepLanguage(stepConfig) {
        if (stepConfig.filePath.endsWith('.py')) {
            return this.getLanguage('python');
        }
        else if (stepConfig.filePath.endsWith('.js') || stepConfig.filePath.endsWith('.ts')) {
            return this.getLanguage('node');
        }
        else if (stepConfig.filePath.endsWith('.rb')) {
            return this.getLanguage('ruby');
        }
        return this.getLanguage('unknown');
    }
    getStepType(stepConfig) {
        if (stepConfig.config.type === 'api') {
            return colors.gray('(API)');
        }
        else if (stepConfig.config.type === 'event') {
            return colors.gray('(Event)');
        }
        else if (stepConfig.config.type === 'cron') {
            return colors.gray('(Cron)');
        }
        else if (stepConfig.config.type === 'noop') {
            return colors.gray('(Noop)');
        }
        return colors.gray('(Unknown)');
    }
    getRelativePath(path) {
        return colors.bold(colors.cyan(this.printer.getRelativePath(path)));
    }
    getStepPath(stepConfig) {
        return this.getRelativePath(stepConfig.filePath);
    }
    printStepUploading(stepConfig) {
        const stepLanguage = this.getStepLanguage(stepConfig);
        const stepType = this.getStepType(stepConfig);
        const stepPath = this.getStepPath(stepConfig);
        const stepTag = this.printer.stepTag;
        this.output.log(stepConfig.filePath, (message) => {
            message.append(`${uploading} ${stepLanguage} ${stepTag} ${stepType} ${stepPath}`);
        });
    }
    printStepUploaded(stepConfig) {
        const stepLanguage = this.getStepLanguage(stepConfig);
        const stepType = this.getStepType(stepConfig);
        const stepPath = this.getStepPath(stepConfig);
        const stepTag = this.printer.stepTag;
        this.output.log(stepConfig.filePath, (message) => {
            message.append(`${uploaded} ${stepLanguage} ${stepTag} ${stepType} ${stepPath}`);
        });
    }
    printRouterUploading(language, routerPath) {
        const stepLanguage = this.getLanguage(language);
        const stepPath = this.getRelativePath(routerPath);
        this.output.log(routerPath, (message) => message.append(`${uploading} ${stepLanguage} ${stepPath}`));
    }
    printRouterUploaded(language, routerPath) {
        const stepLanguage = this.getLanguage(language);
        const stepPath = this.getRelativePath(routerPath);
        this.output.log(routerPath, (message) => message.append(`${uploaded} ${stepLanguage} ${stepPath}`));
    }
    printConfigurationUploading() {
        this.output.log('upload-config', (message) => message.tag('progress').append(`${uploading} Configuration`));
    }
    printConfigurationUploaded() {
        this.output.log('upload-config', (message) => message.tag('success').append(`${uploaded} Configuration`));
    }
}
