// motia/dist/cjs/create-step/getAnswers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStepAnswers = void 0;
const colors_1 = __importDefault(require("colors"));
const inquirer_1 = __importDefault(require("inquirer"));
const types_1 = require("./types");
const getStepAnswers = async () => {
    console.log('\nðŸ“ ', colors_1.default.bold('Create a new Motia step\n'));
    // Basic information prompts
    const basicInfo = await inquirer_1.default.prompt([
        {
            type: 'input',
            name: 'name',
            message: 'Step name:',
            validate: (input) => {
                if (input.length === 0)
                    return 'Name is required';
                if (!/^[a-zA-Z][a-zA-Z0-9-_]*$/.test(input)) {
                    return 'Name must start with a letter and contain only letters, numbers, hyphens, and underscores';
                }
                return true;
            },
        },
        {
            type: 'list',
            name: 'type',
            message: 'Select step type:',
            choices: types_1.STEP_TYPES.map((type) => ({
                name: type.toUpperCase(),
                value: type,
            })),
        },
    ]);
    let answers = { ...basicInfo };
    const languageAnswer = {
        type: 'list',
        name: 'language',
        message: 'Select language:',
        choices: types_1.LANGUAGES.map((lang) => ({
            name: lang.charAt(0).toUpperCase() + lang.slice(1),
            value: lang,
        })),
    };
    // Type-specific configuration prompts
    if (answers.type === 'api') {
        const apiConfig = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'method',
                message: 'HTTP method:',
                choices: types_1.HTTP_METHODS,
            },
            {
                type: 'input',
                name: 'path',
                message: 'API path (e.g. /users):',
                validate: (input) => {
                    if (!input.startsWith('/'))
                        return 'Path must start with /';
                    return true;
                },
            },
        ]);
        answers = { ...answers, ...apiConfig };
    }
    else if (answers.type === 'event') {
        const eventConfig = await inquirer_1.default.prompt([
            languageAnswer,
            {
                type: 'input',
                name: 'subscriptions',
                message: 'Event subscriptions (comma-separated):',
                filter: (input) => input
                    .split(',')
                    .map((s) => s.trim())
                    .filter(Boolean),
            },
        ]);
        answers = { ...answers, ...eventConfig };
    }
    else if (answers.type === 'cron') {
        const cronConfig = await inquirer_1.default.prompt([
            languageAnswer,
            {
                type: 'input',
                name: 'cronExpression',
                message: 'Cron expression:',
                validate: (input) => {
                    if (!input)
                        return 'Cron expression is required';
                    const parts = input.split(' ');
                    if (parts.length !== 5)
                        return 'Invalid cron expression format';
                    return true;
                },
            },
        ]);
        answers = { ...answers, ...cronConfig };
    }
    else if (answers.type === 'noop') {
        const noopConfig = await inquirer_1.default.prompt([
            languageAnswer,
            {
                type: 'input',
                name: 'virtualEmits',
                message: 'Virtual emits (comma-separated):',
                filter: (input) => input
                    .split(',')
                    .map((s) => s.trim())
                    .filter(Boolean),
            },
            {
                type: 'input',
                name: 'virtualSubscribes',
                message: 'Virtual subscribes (comma-separated):',
                filter: (input) => input
                    .split(',')
                    .map((s) => s.trim())
                    .filter(Boolean),
            },
        ]);
        answers = { ...answers, ...noopConfig };
    }
    // Common configuration prompts
    const commonConfig = await inquirer_1.default.prompt([
        {
            type: 'input',
            name: 'flows',
            message: 'Flow names (comma-separated):',
            filter: (input) => input
                .split(',')
                .map((s) => s.trim())
                .filter(Boolean),
            validate: (input) => {
                if (input.length === 0)
                    return 'At least one flow is required';
                return true;
            },
        },
        {
            type: 'input',
            name: 'emits',
            message: 'Events to emit (comma-separated):',
            filter: (input) => input
                .split(',')
                .map((s) => s.trim())
                .filter(Boolean),
        },
        {
            type: 'input',
            name: 'description',
            message: 'Step description:',
        },
        {
            type: 'confirm',
            name: 'createOverride',
            message: 'Create UI component override?',
            default: false,
        },
    ]);
    const nextAnswers = { ...answers, ...commonConfig };
    return {
        ...nextAnswers,
        language: nextAnswers.type === 'api' ? 'typescript' : nextAnswers.language,
    };
};
exports.getStepAnswers = getStepAnswers;
