// motia/dist/esm/cloud/build/builders/archiver.js
import archiver from 'archiver';
import fs from 'fs';
import path from 'path';
export class Archiver {
    constructor(filePath) {
        this.uncompressedSize = 0;
        this.archive = archiver('zip', { zlib: { level: 9 } });
        this.outputStream = fs.createWriteStream(filePath);
        this.archive.pipe(this.outputStream);
    }
    appendDirectory(sourcePath, targetPath) {
        try {
            const stat = fs.statSync(sourcePath);
            if (!stat.isDirectory()) {
                return;
            }
            this.uncompressedSize += this.calculateDirectorySize(sourcePath);
            this.archive.directory(sourcePath, targetPath === '/' ? false : targetPath);
        }
        catch (_error) { }
    }
    calculateDirectorySize(dirPath) {
        let totalSize = 0;
        try {
            const items = fs.readdirSync(dirPath);
            for (const item of items) {
                const fullPath = path.join(dirPath, item);
                try {
                    const stat = fs.statSync(fullPath);
                    if (stat.isDirectory()) {
                        totalSize += this.calculateDirectorySize(fullPath);
                    }
                    else {
                        totalSize += stat.size;
                    }
                }
                catch (_error) { }
            }
        }
        catch (_error) { }
        return totalSize;
    }
    append(stream, filePath) {
        if (typeof stream === 'string') {
            this.uncompressedSize += Buffer.byteLength(stream, 'utf8');
            this.archive.append(stream, { name: filePath });
        }
        else if (Buffer.isBuffer(stream)) {
            this.uncompressedSize += stream.length;
            this.archive.append(stream, { name: filePath });
        }
        else {
            const stats = fs.statSync(stream.path);
            this.uncompressedSize += stats.size;
            this.archive.append(stream, { name: filePath });
        }
    }
    async finalize() {
        return new Promise((resolve, reject) => {
            this.outputStream.on('close', () => {
                resolve({
                    compressedSize: this.archive.pointer(),
                    uncompressedSize: this.uncompressedSize,
                });
            });
            this.outputStream.on('error', reject);
            this.archive.finalize();
        });
    }
}
