// motia/dist/cjs/create-step/templates/ui/overrides.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateOverride = generateOverride;
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
/**
 * Generates a React component override for the step
 */
async function generateOverride(answers) {
    const templatePath = path_1.default.join(__dirname, `${answers.type}.step.txt`);
    const content = await fs_1.promises.readFile(templatePath, 'utf8');
    const replacements = {
        STEP_NAME: answers.name,
        DESCRIPTION: answers.description || '',
        TYPE: answers.type,
        FLOWS: JSON.stringify(answers.flows),
        EMITS: JSON.stringify(answers.emits),
    };
    return Object.entries(replacements).reduce((content, [key, value]) => {
        const regex = new RegExp(`\\{\\{${key}\\}\\}`, 'g');
        return content.replace(regex, String(value));
    }, content);
}
