// motia/dist/cjs/utils/analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapAction = exports.logCliError = exports.identifyUser = exports.disableAnalytics = exports.enableAnalytics = void 0;
const analytics_node_1 = require("@amplitude/analytics-node");
const core_1 = require("@motiadev/core");
const utils_1 = require("@motiadev/core/dist/src/analytics/utils");
const version_1 = require("../version");
const enrichment_plugin_1 = require("./amplitude/enrichment-plugin");
const build_error_1 = require("./errors/build.error");
(0, analytics_node_1.init)('ab2408031a38aa5cb85587a27ecfc69c', {
    logLevel: analytics_node_1.Types.LogLevel.None,
});
const updateOptOutStatus = () => {
    const optOut = !(0, core_1.isAnalyticsEnabled)();
    (0, analytics_node_1.setOptOut)(optOut);
};
updateOptOutStatus();
(0, analytics_node_1.add)(new enrichment_plugin_1.MotiaEnrichmentPlugin());
const enableAnalytics = () => {
    (0, analytics_node_1.setOptOut)(false);
};
exports.enableAnalytics = enableAnalytics;
const disableAnalytics = () => {
    (0, analytics_node_1.setOptOut)(true);
};
exports.disableAnalytics = disableAnalytics;
const identifyUser = () => {
    try {
        const identifyObj = new analytics_node_1.Identify();
        identifyObj.postInsert('project_id', (0, utils_1.getProjectName)(process.cwd()));
        identifyObj.postInsert('motia_version', version_1.version || 'unknown');
        identifyObj.postInsert('project_version', process.env.npm_package_version || 'unknown');
        (0, analytics_node_1.identify)(identifyObj, {
            user_id: (0, core_1.getUserIdentifier)(),
        });
    }
    catch (error) {
        // Silently fail
    }
};
exports.identifyUser = identifyUser;
const logCliError = (command, error) => {
    try {
        const cause = error.cause instanceof build_error_1.BuildError ? error.cause : error;
        const errorMessage = cause?.message || 'Unknown error';
        const errorType = cause?.constructor?.name || 'Unknown error type';
        const errorStack = cause?.stack ? cause.stack.split('\n').slice(0, 10).join('\n') : undefined;
        const truncatedMessage = errorMessage.length > 500 ? `${errorMessage.substring(0, 500)}...` : errorMessage;
        (0, core_1.trackEvent)('cli_command_error', {
            command,
            error_message: truncatedMessage,
            error_type: errorType,
            ...(errorStack && { error_stack: errorStack }),
        });
    }
    catch (logError) {
        // Silently fail to not disrupt CLI operations
    }
};
exports.logCliError = logCliError;
const getCommandNameFromArgs = () => {
    const args = process.argv.slice(2);
    const commandParts = [];
    for (let i = 0; i < args.length && i < 3; i++) {
        const arg = args[i];
        if (!arg.startsWith('-') && !arg.startsWith('--')) {
            commandParts.push(arg);
        }
        else {
            break;
        }
    }
    return commandParts.join(' ') || 'unknown';
};
const wrapAction = (action) => {
    return (async (...args) => {
        try {
            return await action(...args);
        }
        catch (error) {
            const commandName = getCommandNameFromArgs();
            (0, exports.logCliError)(commandName, error);
            await (0, analytics_node_1.flush)().promise.catch(() => {
                // Silently fail
            });
            throw error;
        }
    });
};
exports.wrapAction = wrapAction;
