// motia/dist/cjs/utils/internal-logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.internalLogger = void 0;
const colors_1 = __importDefault(require("colors"));
const infoTag = colors_1.default.blue('➜ [INFO]');
const warnTag = colors_1.default.yellow('➜ [WARN]');
const errorTag = colors_1.default.red('✘ [ERROR]');
const colorMap = {
    cyan: colors_1.default.cyan,
    yellow: colors_1.default.yellow,
    red: colors_1.default.red,
};
const extraTag = (color, extra) => {
    const colorFn = colorMap[color];
    return extra ? colors_1.default.bold(colorFn(extra)) : '';
};
exports.internalLogger = {
    info: (message, extra) => {
        console.log(`${infoTag} ${message} ${extraTag('cyan', extra)}`);
    },
    warn: (message, extra) => {
        console.log(`${warnTag} ${message} ${extraTag('yellow', extra)}`);
    },
    error: (message, extra) => {
        console.log(`${errorTag} ${message} ${extraTag('red', extra)}`);
    },
};
