// @motiadev/workbench/dist/src/App.js
import { jsx as _jsx } from "react/jsx-runtime";
import { memo, useEffect, useMemo } from 'react';
import { FlowPage } from './components/flow/flow-page';
import { FlowTabMenuItem } from './components/flow/flow-tab-menu-item';
import { registerPluginTabs } from './lib/plugins';
import { getViewModeFromURL } from './lib/utils';
import { ProjectViewMode } from './project-view-mode';
import { TabLocation, useAppTabsStore } from './stores/use-app-tabs-store';
import { SystemViewMode } from './system-view-mode';
const TAB_IDS = {
    FLOW: 'flow',
    LOGS: 'logs',
};
const topTabs = [
    {
        id: TAB_IDS.FLOW,
        tabLabel: FlowTabMenuItem,
        content: FlowPage,
    },
];
export const App = memo(() => {
    const setTabs = useAppTabsStore((state) => state.setTabs);
    const addTab = useAppTabsStore((state) => state.addTab);
    useEffect(() => {
        const timeout = setTimeout(() => {
            setTabs(TabLocation.TOP, topTabs);
            registerPluginTabs(addTab);
        }, 10);
        return () => clearTimeout(timeout);
    }, [setTabs, addTab]);
    const viewMode = useMemo(getViewModeFromURL, []);
    const ViewComponent = viewMode === 'project' ? ProjectViewMode : SystemViewMode;
    return _jsx(ViewComponent, {});
});
App.displayName = 'App';
