// motia/dist/cjs/cloud/build/validations/__tests__/api-endpoints.validator.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api_endpoints_validator_1 = require("../api-endpoints.validator");
describe('apiEndpointsValidator', () => {
    const createMockBuilder = (stepsConfig) => {
        return {
            projectDir: '/project',
            stepsConfig,
            streamsConfig: {},
            routersConfig: {},
            stepUncompressedSizes: new Map(),
            stepCompressedSizes: new Map(),
            routerUncompressedSizes: new Map(),
            routerCompressedSizes: new Map(),
            modulegraphInstalled: false,
            registerBuilder: jest.fn(),
            registerStateStream: jest.fn(),
            registerStep: jest.fn(),
            recordStepSize: jest.fn(),
            buildStep: jest.fn(),
            buildApiSteps: jest.fn(),
        };
    };
    const createMockApiStep = (name, filePath, method, path) => ({
        type: 'node',
        entrypointPath: filePath,
        config: { name, type: 'api', method, path },
        filePath,
    });
    it('should return no errors when all endpoints are unique', () => {
        const builder = createMockBuilder({
            '/step1.ts': createMockApiStep('step1', '/project/step1.ts', 'GET', '/users'),
            '/step2.ts': createMockApiStep('step2', '/project/step2.ts', 'POST', '/users'),
            '/step3.ts': createMockApiStep('step3', '/project/step3.ts', 'GET', '/posts'),
        });
        const result = (0, api_endpoints_validator_1.apiEndpointsValidator)(builder);
        expect(result.errors).toHaveLength(0);
        expect(result.warnings).toHaveLength(0);
    });
    it('should return error when endpoints conflict', () => {
        const builder = createMockBuilder({
            '/step1.ts': createMockApiStep('step1', '/project/step1.ts', 'GET', '/users'),
            '/step2.ts': createMockApiStep('step2', '/project/step2.ts', 'GET', '/users'),
        });
        const result = (0, api_endpoints_validator_1.apiEndpointsValidator)(builder);
        expect(result.errors).toHaveLength(1);
        expect(result.errors[0].message).toContain('Endpoint conflict');
        expect(result.errors[0].message).toContain('GET /users');
    });
    it('should skip non-api steps', () => {
        const builder = createMockBuilder({
            '/step1.ts': {
                type: 'node',
                entrypointPath: '/project/step1.ts',
                config: { name: 'step1', type: 'event' },
                filePath: '/project/step1.ts',
            },
        });
        const result = (0, api_endpoints_validator_1.apiEndpointsValidator)(builder);
        expect(result.errors).toHaveLength(0);
    });
    it('should allow same path with different methods', () => {
        const builder = createMockBuilder({
            '/step1.ts': createMockApiStep('step1', '/project/step1.ts', 'GET', '/users'),
            '/step2.ts': createMockApiStep('step2', '/project/step2.ts', 'POST', '/users'),
            '/step3.ts': createMockApiStep('step3', '/project/step3.ts', 'DELETE', '/users'),
        });
        const result = (0, api_endpoints_validator_1.apiEndpointsValidator)(builder);
        expect(result.errors).toHaveLength(0);
    });
});
