// motia/dist/cjs/cloud/build/validations/__tests__/router-bundle-sizes.validator.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const router_bundle_sizes_validator_1 = require("../router-bundle-sizes.validator");
describe('routerBundleSizesValidator', () => {
    const createMockBuilder = (routerSizes = new Map()) => {
        return {
            projectDir: '/project',
            stepsConfig: {
                '/step1.ts': {
                    type: 'node',
                    entrypointPath: '/project/step1.ts',
                    config: { name: 'step1', type: 'api' },
                    filePath: '/project/step1.ts',
                },
            },
            streamsConfig: {},
            routersConfig: {},
            stepUncompressedSizes: new Map(),
            stepCompressedSizes: new Map(),
            routerUncompressedSizes: routerSizes,
            routerCompressedSizes: new Map(),
            modulegraphInstalled: false,
            registerBuilder: jest.fn(),
            registerStateStream: jest.fn(),
            registerStep: jest.fn(),
            recordStepSize: jest.fn(),
            buildStep: jest.fn(),
            buildApiSteps: jest.fn(),
        };
    };
    it('should return no errors when router size is within limits', () => {
        const routerSizes = new Map([['node', 100 * constants_1.BYTES_TO_MB]]);
        const builder = createMockBuilder(routerSizes);
        const result = (0, router_bundle_sizes_validator_1.routerBundleSizesValidator)(builder);
        expect(result.errors).toHaveLength(0);
        expect(result.warnings).toHaveLength(0);
    });
    it('should return error when router size exceeds limit', () => {
        const routerSizes = new Map([['node', (constants_1.BUNDLE_SIZE_LIMITS.ROUTER_MAX_MB + 1) * constants_1.BYTES_TO_MB]]);
        const builder = createMockBuilder(routerSizes);
        const result = (0, router_bundle_sizes_validator_1.routerBundleSizesValidator)(builder);
        expect(result.errors).toHaveLength(1);
        expect(result.errors[0].message).toContain('exceeds');
        expect(result.errors[0].message).toContain('150MB');
    });
    it('should handle multiple routers', () => {
        const routerSizes = new Map([
            ['node', 100 * constants_1.BYTES_TO_MB],
            ['python', 120 * constants_1.BYTES_TO_MB],
        ]);
        const builder = createMockBuilder(routerSizes);
        const result = (0, router_bundle_sizes_validator_1.routerBundleSizesValidator)(builder);
        expect(result.errors).toHaveLength(0);
    });
    it('should return errors for multiple oversized routers', () => {
        const oversizedValue = (constants_1.BUNDLE_SIZE_LIMITS.ROUTER_MAX_MB + 10) * constants_1.BYTES_TO_MB;
        const routerSizes = new Map([
            ['node', oversizedValue],
            ['python', oversizedValue],
        ]);
        const builder = createMockBuilder(routerSizes);
        const result = (0, router_bundle_sizes_validator_1.routerBundleSizesValidator)(builder);
        expect(result.errors).toHaveLength(2);
    });
    it('should handle empty router sizes', () => {
        const builder = createMockBuilder(new Map());
        const result = (0, router_bundle_sizes_validator_1.routerBundleSizesValidator)(builder);
        expect(result.errors).toHaveLength(0);
    });
});
