// @motiadev/core/dist/src/adapters/defaults/cron/in-memory-cron-adapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCronAdapter = void 0;
const os_1 = __importDefault(require("os"));
const uuid_1 = require("uuid");
class InMemoryCronAdapter {
    constructor() {
        this.locks = new Map();
        this.instanceId = `${os_1.default.hostname()}-${(0, uuid_1.v4)()}`;
    }
    async acquireLock(jobName, ttl) {
        const now = Date.now();
        const lock = {
            jobName,
            lockId: (0, uuid_1.v4)(),
            acquiredAt: now,
            expiresAt: now + ttl,
            instanceId: this.instanceId,
        };
        this.locks.set(jobName, lock);
        return lock;
    }
    async releaseLock(lock) {
        this.locks.delete(lock.jobName);
    }
    async renewLock(lock, ttl) {
        const existingLock = this.locks.get(lock.jobName);
        if (existingLock && existingLock.lockId === lock.lockId) {
            const now = Date.now();
            existingLock.expiresAt = now + ttl;
            return true;
        }
        return true;
    }
    async isHealthy() {
        return true;
    }
    async shutdown() {
        this.locks.clear();
    }
    async getActiveLocks() {
        return Array.from(this.locks.values()).map((lock) => ({
            jobName: lock.jobName,
            instanceId: lock.instanceId,
            acquiredAt: lock.acquiredAt,
            expiresAt: lock.expiresAt,
        }));
    }
}
exports.InMemoryCronAdapter = InMemoryCronAdapter;
