// @motiadev/core/dist/src/printer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoPrinter = exports.Printer = void 0;
const colors_1 = __importDefault(require("colors"));
const path_1 = __importDefault(require("path"));
const guards_1 = require("./guards");
const stepTag = colors_1.default.bold(colors_1.default.magenta('Step'));
const flowTag = colors_1.default.bold(colors_1.default.blue('Flow'));
const streamTag = colors_1.default.bold(colors_1.default.green('Stream'));
const registered = colors_1.default.green('➜ [REGISTERED]');
const building = colors_1.default.yellow('⚡ [BUILDING]');
const built = colors_1.default.green('✓ [BUILT]');
const updated = colors_1.default.yellow('➜ [UPDATED]');
const removed = colors_1.default.red('➜ [REMOVED]');
const invalidEmit = colors_1.default.red('➜ [INVALID EMIT]');
const error = colors_1.default.red('[ERROR]');
const warning = colors_1.default.yellow('[WARNING]');
const warnIcon = colors_1.default.yellow('⚠');
const infoIcon = colors_1.default.blue('ℹ');
const errorIcon = colors_1.default.red('✖');
class Printer {
    constructor(baseDir) {
        this.baseDir = baseDir;
        this.stepTag = stepTag;
        this.flowTag = flowTag;
        this.registered = registered;
        this.building = building;
        this.built = built;
        this.updated = updated;
        this.removed = removed;
    }
    printEventInputValidationError(emit, details) {
        const emitPath = colors_1.default.bold(colors_1.default.cyan(`Emit ${emit.topic}`));
        console.log(`${warnIcon} ${warning} ${emitPath} validation issues:`);
        const hasAny = details.missingFields?.length || details.extraFields?.length || details.typeMismatches?.length;
        if (!hasAny) {
            console.log(`${colors_1.default.yellow('│')} No issues found.`);
            console.log(`${colors_1.default.yellow('└─')} Validation passed.`);
            return;
        }
        if (details.missingFields?.length) {
            console.log(`${colors_1.default.yellow('│')} ${colors_1.default.yellow(`⚠ Missing fields: ${details.missingFields.join(', ')}`)}`);
        }
        if (details.extraFields?.length) {
            console.log(`${colors_1.default.yellow('│')} ${colors_1.default.yellow(`⚠ Extra fields: ${details.extraFields.join(', ')}`)}`);
        }
        if (details.typeMismatches?.length) {
            console.log(`${colors_1.default.yellow('│')} ${colors_1.default.yellow(`⚠ Type mismatches: ${details.typeMismatches.join(', ')}`)}`);
        }
        console.log(`${colors_1.default.yellow('└─')} ${colors_1.default.yellow('Payload does not match schema.')}`);
    }
    printInvalidEmit(step, emit) {
        console.log(`${invalidEmit} ${stepTag} ${this.getStepType(step)} ${this.getStepPath(step)} tried to emit an event not defined in the step config: ${colors_1.default.yellow(emit)}`);
    }
    printStepCreated(step) {
        console.log(`${registered} ${stepTag} ${this.getStepType(step)} ${this.getStepPath(step)} registered`);
    }
    printStepUpdated(step) {
        console.log(`${updated} ${stepTag} ${this.getStepType(step)} ${this.getStepPath(step)} updated`);
    }
    printStepRemoved(step) {
        console.log(`${removed} ${stepTag} ${this.getStepType(step)} ${this.getStepPath(step)} removed`);
    }
    printFlowCreated(flowName) {
        console.log(`${registered} ${flowTag} ${colors_1.default.bold(colors_1.default.cyan(flowName))} registered`);
    }
    printFlowUpdated(flowName) {
        console.log(`${updated} ${flowTag} ${colors_1.default.bold(colors_1.default.cyan(flowName))} updated`);
    }
    printFlowRemoved(flowName) {
        console.log(`${removed} ${flowTag} ${colors_1.default.bold(colors_1.default.cyan(flowName))} removed`);
    }
    printStreamCreated(stream) {
        console.log(`${registered} ${streamTag} ${this.getStreamPath(stream)} registered`);
    }
    printStreamUpdated(stream) {
        console.log(`${updated} ${streamTag} ${this.getStreamPath(stream)} updated`);
    }
    printStreamRemoved(stream) {
        console.log(`${removed} ${streamTag} ${this.getStreamPath(stream)} removed`);
    }
    printInvalidEmitConfiguration(step, emit) {
        console.log(`${warnIcon} ${warning} ${stepTag} ${this.getStepType(step)} ${this.getStepPath(step)} emits to ${colors_1.default.yellow(emit)}, but there is no subscriber defined`);
    }
    printInvalidSchema(topic, step) {
        console.log(`${error} Topic ${colors_1.default.bold(colors_1.default.blue(topic))} has incompatible schemas in the following steps:`);
        step.forEach((step) => {
            console.log(`${colors_1.default.red('  ✖')} ${this.getStepPath(step)}`);
        });
    }
    printValidationError(stepPath, validationError) {
        const relativePath = this.getRelativePath(stepPath);
        console.log(`${error} ${colors_1.default.bold(colors_1.default.cyan(relativePath))}`);
        validationError.errors?.forEach((error) => {
            if (error.path) {
                console.log(`${colors_1.default.red('│')} ${colors_1.default.yellow(`✖ ${error.path}`)}: ${error.message}`);
            }
            else {
                console.log(`${colors_1.default.red('│')} ${colors_1.default.yellow('✖')} ${error.message}`);
            }
        });
        console.log(`${colors_1.default.red('└─')} ${colors_1.default.red(validationError.error)}  `);
    }
    getRelativePath(filePath) {
        return path_1.default.relative(this.baseDir, filePath);
    }
    getStepType(step) {
        if ((0, guards_1.isApiStep)(step))
            return colors_1.default.gray('(API)');
        if ((0, guards_1.isEventStep)(step))
            return colors_1.default.gray('(Event)');
        if ((0, guards_1.isCronStep)(step))
            return colors_1.default.gray('(Cron)');
        if ((0, guards_1.isNoopStep)(step))
            return colors_1.default.gray('(Noop)');
        return colors_1.default.gray('(Unknown)');
    }
    getStepPath(step) {
        const stepPath = this.getRelativePath(step.filePath);
        return colors_1.default.bold(colors_1.default.cyan(stepPath));
    }
    getStreamPath(stream) {
        const streamPath = this.getRelativePath(stream.filePath);
        return colors_1.default.bold(colors_1.default.magenta(streamPath));
    }
    printPluginLog(message) {
        const pluginTag = colors_1.default.bold(colors_1.default.cyan('[motia-plugins]'));
        console.log(`${infoIcon} ${pluginTag} ${message}`);
    }
    printPluginWarn(message) {
        const pluginTag = colors_1.default.bold(colors_1.default.cyan('[motia-plugins]'));
        console.warn(`${warnIcon} ${pluginTag} ${colors_1.default.yellow(message)}`);
    }
    printPluginError(message, ...args) {
        const pluginTag = colors_1.default.bold(colors_1.default.cyan('[motia-plugins]'));
        console.error(`${errorIcon} ${pluginTag} ${colors_1.default.red(message)}`, ...args);
    }
}
exports.Printer = Printer;
class NoPrinter extends Printer {
    constructor() {
        super('');
    }
    printEventInputValidationError() { }
    printInvalidEmit() { }
    printStepCreated() { }
    printStepUpdated() { }
    printStepRemoved() { }
    printFlowCreated() { }
    printFlowUpdated() { }
    printFlowRemoved() { }
    printStepType() { }
    printStepPath() { }
    printStreamCreated() { }
    printStreamUpdated() { }
    printStreamRemoved() { }
    printPluginLog() { }
    printPluginWarn() { }
    printPluginError() { }
}
exports.NoPrinter = NoPrinter;
