// motia/dist/cjs/cloud/cli/deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const build_validation_1 = require("../build/build-validation");
const cli_1 = require("../cli");
const config_utils_1 = require("../config-utils");
const build_1 = require("../new-deployment/build");
const cloud_api_1 = require("../new-deployment/cloud-api");
const deploy_1 = require("../new-deployment/deploy");
const cli_listener_1 = require("../new-deployment/listeners/cli-listener");
const upload_artifacts_1 = require("../new-deployment/upload-artifacts");
const load_env_data_1 = require("../new-deployment/utils/load-env-data");
cli_1.cloudCli
    .command('deploy')
    .description('Deploy a new version to Motia Cloud')
    .requiredOption('-k, --api-key <key>', 'The API key for authentication', process.env.MOTIA_API_KEY)
    .requiredOption('-v, --version-name <version>', 'The version to deploy')
    .option('-p, --project-id <id>', 'Project ID (Deprecated)')
    .option('-n, --project-name <name>', 'Project name (used when creating a new project)')
    .option('-s, --environment-id <id>', 'Environment ID', process.env.MOTIA_ENVIRONMENT_ID)
    .option('--environment-name <name>', 'Environment name')
    .option('-e, --env-file <path>', 'Path to environment file')
    .option('-d, --version-description <description>', 'The description of the version')
    .option('-c, --ci', 'CI mode', process.env.CI)
    .action((0, config_utils_1.handler)(async (arg, context) => {
    const listener = new cli_listener_1.CliListener(context);
    const builder = await (0, build_1.build)(listener);
    const isValid = (0, build_validation_1.buildValidation)(builder, listener);
    if (!isValid) {
        process.exit(1);
    }
    context.log('build-completed', (message) => message.tag('success').append('Build completed'));
    context.log('creating-deployment', (message) => message.tag('progress').append('Creating deployment...'));
    const deployment = await cloud_api_1.cloudApi
        .createDeployment({
        apiKey: arg.apiKey,
        projectName: arg.projectName,
        environmentId: arg.environmentId,
        environmentName: arg.environmentName,
        versionName: arg.versionName,
        versionDescription: arg.versionDescription,
    })
        .catch((error) => {
        context.log('creating-deployment', (message) => message.tag('failed').append('Failed to create deployment'));
        throw error;
    });
    context.log('creating-deployment', (message) => message.tag('success').append('Deployment created'));
    context.log('uploading-artifacts', (message) => message.tag('progress').append('Uploading artifacts...'));
    await (0, upload_artifacts_1.uploadArtifacts)(builder, deployment.deploymentToken, listener);
    context.log('uploading-artifacts', (message) => message.tag('success').append('Artifacts uploaded'));
    context.log('starting-deployment', (message) => message.tag('progress').append('Starting deployment...'));
    await (0, deploy_1.deploy)({
        envVars: (0, load_env_data_1.loadEnvData)(arg.envFile, context),
        deploymentId: deployment.deploymentId,
        deploymentToken: deployment.deploymentToken,
        builder,
        listener,
        context,
        ci: arg.ci,
    });
    context.exit(0);
}));
