// @motiadev/core/dist/src/adapters/defaults/stream/file-stream-adapter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileStreamAdapter = void 0;
const fs_1 = __importDefault(require("fs"));
const path = __importStar(require("path"));
const stream_adapter_interface_1 = require("../../interfaces/stream-adapter.interface");
class FileStreamAdapter extends stream_adapter_interface_1.StreamAdapter {
    constructor(filePath, streamName, motiaFileStoragePath = '.motia') {
        super(streamName);
        this.streamsDir = path.join(filePath, motiaFileStoragePath, 'streams');
        this.filePath = path.join(this.streamsDir, `${streamName}.stream.json`);
    }
    init() {
        try {
            fs_1.default.realpathSync(this.streamsDir);
        }
        catch {
            fs_1.default.mkdirSync(this.streamsDir, { recursive: true });
        }
        try {
            fs_1.default.readFileSync(this.filePath, 'utf-8');
        }
        catch {
            fs_1.default.writeFileSync(this.filePath, JSON.stringify({}), 'utf-8');
        }
    }
    async getGroup(groupId) {
        const data = this._readFile();
        const prefix = this._makeKey(groupId, '');
        return Object.entries(data)
            .filter(([key]) => key.startsWith(prefix))
            .map(([, value]) => JSON.parse(value));
    }
    async get(groupId, key) {
        const data = this._readFile();
        const fullKey = this._makeKey(groupId, key);
        return data[fullKey] ? JSON.parse(data[fullKey]) : null;
    }
    async set(groupId, id, value) {
        const data = this._readFile();
        const key = this._makeKey(groupId, id);
        data[key] = JSON.stringify(value);
        this._writeFile(data);
        return { ...value, id };
    }
    async delete(groupId, id) {
        const data = this._readFile();
        const key = this._makeKey(groupId, id);
        const value = await this.get(groupId, id);
        if (value) {
            delete data[key];
            this._writeFile(data);
        }
        return value;
    }
    async clear(groupId) {
        const data = this._readFile();
        const pattern = this._makeKey(groupId, '');
        for (const key in data) {
            if (key.startsWith(pattern)) {
                delete data[key];
            }
        }
        this._writeFile(data);
    }
    _makeKey(groupId, id) {
        return `${groupId}:${id}`;
    }
    _readFile() {
        try {
            const content = fs_1.default.readFileSync(this.filePath, 'utf-8');
            return JSON.parse(content);
        }
        catch (error) {
            this.init();
            return {};
        }
    }
    _writeFile(data) {
        try {
            fs_1.default.writeFileSync(this.filePath, JSON.stringify(data, null, 2), 'utf-8');
        }
        catch (error) {
            this.init();
            fs_1.default.writeFileSync(this.filePath, JSON.stringify(data, null, 2), 'utf-8');
        }
    }
}
exports.FileStreamAdapter = FileStreamAdapter;
