// motia/dist/cjs/cloud/build/builders/python/python-data/__tests__/resolve-dep-name.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const python_errors_1 = require("../python-errors");
const resolve_dep_names_1 = require("../resolve-dep-names");
const sitePackagesDir = path_1.default.join(__dirname, 'site-packages');
describe('resolveDepName', () => {
    test('resolves dependency names correctly', () => {
        const depNames = ['httpx'];
        const result = (0, resolve_dep_names_1.resolveDepNames)(depNames, sitePackagesDir);
        expect(result).toEqual({ httpx: 'httpx' });
    });
    test('resolves multiple dependency names correctly', () => {
        const depNames = ['httpx', 'scikit-learn', 'opencv-python'];
        const result = (0, resolve_dep_names_1.resolveDepNames)(depNames, sitePackagesDir);
        expect(result).toEqual({
            httpx: 'httpx',
            sklearn: 'scikit-learn',
            cv2: 'opencv-python',
        });
    });
    test('resolves pymongo correctly', () => {
        const depNames = ['pymongo'];
        const result = (0, resolve_dep_names_1.resolveDepNames)(depNames, sitePackagesDir);
        expect(result).toEqual({
            pymongo: 'pymongo',
            bson: 'pymongo',
            gridfs: 'pymongo',
        });
    });
    test('should throw an error if the dependency is not found', () => {
        const depNames = ['httpx', 'scikit-learn', 'opencv-python', 'pydantic'];
        expect(() => (0, resolve_dep_names_1.resolveDepNames)(depNames, sitePackagesDir)).toThrow(new python_errors_1.PythonError('Could not find dependency name in site-packages: pydantic', 'pydantic'));
    });
});
