// motia/dist/cjs/dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dev = void 0;
const analytics_node_1 = require("@amplitude/analytics-node");
const adapter_redis_state_1 = require("@motiadev/adapter-redis-state");
const adapter_redis_streams_1 = require("@motiadev/adapter-redis-streams");
const core_1 = require("@motiadev/core");
const endpoints_1 = require("./cloud/endpoints");
const constants_1 = require("./constants");
const dev_watchers_1 = require("./dev-watchers");
const generate_locked_data_1 = require("./generate-locked-data");
const load_motia_config_1 = require("./load-motia-config");
const plugins_1 = require("./plugins");
const redis_memory_manager_1 = require("./redis-memory-manager");
const activate_python_env_1 = require("./utils/activate-python-env");
const analytics_1 = require("./utils/analytics");
const version_1 = require("./version");
process.env.VITE_CJS_IGNORE_WARNING = 'true';
require('ts-node').register({
    transpileOnly: true,
    compilerOptions: { module: 'commonjs' },
});
const dev = async (port, hostname, disableVerbose, enableMermaid, motiaFileStorageDir) => {
    const baseDir = process.cwd();
    const isVerbose = !disableVerbose;
    (0, analytics_1.identifyUser)();
    const stepFiles = (0, generate_locked_data_1.getStepFiles)(baseDir);
    const hasPythonFiles = stepFiles.some((file) => file.endsWith('.py'));
    (0, core_1.trackEvent)('dev_server_started', {
        port,
        verbose_mode: isVerbose,
        mermaid_enabled: enableMermaid,
        has_python_files: hasPythonFiles,
        total_step_files: stepFiles.length,
        project_name: (0, core_1.getProjectIdentifier)(baseDir),
    });
    if (hasPythonFiles) {
        (0, activate_python_env_1.activatePythonVenv)({ baseDir, isVerbose });
        (0, core_1.trackEvent)('python_environment_activated');
    }
    const motiaFileStoragePath = motiaFileStorageDir || '.motia';
    const appConfig = await (0, load_motia_config_1.loadMotiaConfig)(baseDir);
    const redisClient = await (0, redis_memory_manager_1.instanceRedisMemoryServer)(motiaFileStoragePath, true);
    const adapters = {
        eventAdapter: appConfig.adapters?.events || new core_1.DefaultQueueEventAdapter(),
        cronAdapter: appConfig.adapters?.cron || new core_1.DefaultCronAdapter(),
        streamAdapter: appConfig.adapters?.streams || new adapter_redis_streams_1.RedisStreamAdapterManager(redisClient),
    };
    const lockedData = await (0, generate_locked_data_1.generateLockedData)({
        projectDir: baseDir,
        streamAdapter: adapters.streamAdapter,
        redisClient,
    });
    const state = appConfig.adapters?.state || new adapter_redis_state_1.RedisStateAdapter(redisClient);
    const config = { isVerbose };
    const motiaServer = (0, core_1.createServer)(lockedData, state, config, adapters, appConfig.app);
    const watcher = (0, dev_watchers_1.createDevWatchers)(lockedData, motiaServer, motiaServer.motiaEventManager, motiaServer.cronManager);
    const plugins = await (0, plugins_1.processPlugins)(motiaServer);
    // Initialize mermaid generator
    if (enableMermaid) {
        const mermaidGenerator = (0, core_1.createMermaidGenerator)(baseDir);
        mermaidGenerator.initialize(lockedData);
        (0, core_1.trackEvent)('mermaid_generator_initialized');
    }
    (0, endpoints_1.deployEndpoints)(motiaServer, lockedData);
    motiaServer.app.get('/__motia', (_, res) => {
        const meta = {
            version: version_1.version,
            isDev: true,
            isTutorialDisabled: constants_1.isTutorialDisabled,
            workbenchBase: constants_1.workbenchBase,
        };
        res //
            .header('Access-Control-Allow-Origin', '*')
            .header('Access-Control-Allow-Private-Network', 'true')
            .status(200)
            .json(meta);
    });
    (0, core_1.trackEvent)('dev_server_ready', {
        port,
        flows_count: lockedData.flows?.length || 0,
        steps_count: lockedData.activeSteps?.length || 0,
        flows: Object.keys(lockedData.flows || {}),
        steps: lockedData.activeSteps.map((step) => step.config.name),
        streams: Object.keys(lockedData.getStreams() || {}),
        runtime_version: version_1.version,
        environment: process.env.NODE_ENV || 'development',
    });
    const { applyMiddleware } = process.env.__MOTIA_DEV_MODE__
        ? require('@motiadev/workbench/middleware')
        : require('@motiadev/workbench/dist/middleware');
    await applyMiddleware({
        app: motiaServer.app,
        port,
        workbenchBase: constants_1.workbenchBase,
        plugins: plugins.flatMap((item) => item.workbench),
    });
    motiaServer.server.listen(port, hostname);
    console.log('ðŸš€ Server ready and listening on port', port);
    console.log(`ðŸ”— Open http://localhost:${port}${constants_1.workbenchBase} to open workbench ðŸ› ï¸`);
    process.on('SIGTERM', async () => {
        (0, core_1.trackEvent)('dev_server_shutdown', { reason: 'SIGTERM' });
        motiaServer.server.close();
        await watcher.stop();
        await (0, redis_memory_manager_1.stopRedisMemoryServer)();
        await (0, analytics_node_1.flush)().promise;
        process.exit(0);
    });
    process.on('SIGINT', async () => {
        (0, core_1.trackEvent)('dev_server_shutdown', { reason: 'SIGINT' });
        motiaServer.server.close();
        await watcher.stop();
        await (0, redis_memory_manager_1.stopRedisMemoryServer)();
        await (0, analytics_node_1.flush)().promise;
        process.exit(0);
    });
};
exports.dev = dev;
