// motia/dist/cjs/start.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.start = void 0;
const adapter_redis_state_1 = require("@motiadev/adapter-redis-state");
const adapter_redis_streams_1 = require("@motiadev/adapter-redis-streams");
const core_1 = require("@motiadev/core");
const path_1 = __importDefault(require("path"));
const constants_1 = require("./constants");
const generate_locked_data_1 = require("./generate-locked-data");
const load_motia_config_1 = require("./load-motia-config");
const plugins_1 = require("./plugins");
const redis_memory_manager_1 = require("./redis-memory-manager");
const activate_python_env_1 = require("./utils/activate-python-env");
const version_1 = require("./version");
require('ts-node').register({
    transpileOnly: true,
    compilerOptions: { module: 'commonjs' },
});
const start = async (port, hostname, disableVerbose, motiaFileStorageDir) => {
    const baseDir = process.cwd();
    const isVerbose = !disableVerbose;
    const stepFiles = (0, generate_locked_data_1.getStepFiles)(baseDir);
    const hasPythonFiles = stepFiles.some((file) => file.endsWith('.py'));
    if (hasPythonFiles) {
        console.log('âš™ï¸ Activating Python environment...');
        (0, activate_python_env_1.activatePythonVenv)({ baseDir, isVerbose });
    }
    const motiaFileStoragePath = motiaFileStorageDir || '.motia';
    const dotMotia = path_1.default.join(baseDir, motiaFileStoragePath);
    const appConfig = await (0, load_motia_config_1.loadMotiaConfig)(baseDir);
    const redisClient = await (0, redis_memory_manager_1.instanceRedisMemoryServer)(dotMotia);
    const adapters = {
        eventAdapter: appConfig.adapters?.events || new core_1.DefaultQueueEventAdapter(),
        cronAdapter: appConfig.adapters?.cron || new core_1.DefaultCronAdapter(),
        streamAdapter: appConfig.adapters?.streams || new adapter_redis_streams_1.RedisStreamAdapterManager(redisClient),
    };
    const lockedData = await (0, generate_locked_data_1.generateLockedData)({
        projectDir: baseDir,
        streamAdapter: adapters.streamAdapter,
        redisClient,
    });
    const state = appConfig.adapters?.state || new adapter_redis_state_1.RedisStateAdapter(redisClient);
    const config = { isVerbose, isDev: false, version: version_1.version };
    const motiaServer = (0, core_1.createServer)(lockedData, state, config, adapters, appConfig.app);
    const plugins = await (0, plugins_1.processPlugins)(motiaServer);
    if (!process.env.MOTIA_DOCKER_DISABLE_WORKBENCH) {
        const { applyMiddleware } = require('@motiadev/workbench/dist/middleware');
        await applyMiddleware({
            app: motiaServer.app,
            port,
            workbenchBase: constants_1.workbenchBase,
            plugins: plugins.flatMap((item) => item.workbench),
        });
    }
    motiaServer.server.listen(port, hostname);
    console.log('ðŸš€ Server ready and listening on port', port);
    console.log(`ðŸ”— Open http://${hostname}:${port}${constants_1.workbenchBase} to open workbench ðŸ› ï¸`);
    process.on('SIGTERM', async () => {
        motiaServer.server.close();
        await (0, redis_memory_manager_1.stopRedisMemoryServer)();
        process.exit(0);
    });
    process.on('SIGINT', async () => {
        motiaServer.server.close();
        await (0, redis_memory_manager_1.stopRedisMemoryServer)();
        process.exit(0);
    });
};
exports.start = start;
