// @motiadev/core/dist/src/node/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
class Logger {
    constructor(traceId, flows, sender) {
        this.traceId = traceId;
        this.flows = flows;
        this.sender = sender;
        this.listeners = [];
    }
    _log(level, message, args) {
        const argsCopy = args ? { ...args } : {};
        if (argsCopy.error && argsCopy.error instanceof Error) {
            argsCopy.error = {
                ...argsCopy.error,
                message: argsCopy.error.message,
                stack: argsCopy.error.stack,
                name: argsCopy.error.name,
            };
        }
        const logEntry = {
            ...argsCopy,
            level,
            time: Date.now(),
            traceId: this.traceId,
            flows: this.flows,
            msg: message,
        };
        this.sender.sendNoWait('log', logEntry);
        this.listeners.forEach((listener) => listener(level, message, args));
    }
    info(message, args) {
        this._log('info', message, args);
    }
    error(message, args) {
        this._log('error', message, args);
    }
    debug(message, args) {
        this._log('debug', message, args);
    }
    warn(message, args) {
        this._log('warn', message, args);
    }
    addListener(listener) {
        this.listeners.push(listener);
    }
}
exports.Logger = Logger;
