// @motiadev/core/dist/src/node/rpc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcSender = void 0;
/// <reference types="node" />
const crypto_1 = __importDefault(require("crypto"));
class RpcSender {
    constructor(process) {
        this.process = process;
        this.pendingRequests = {};
    }
    async close() {
        const outstandingRequests = Object.values(this.pendingRequests);
        if (outstandingRequests.length > 0) {
            console.error('Process ended while there are some promises outstanding');
            this.process.exit(1);
        }
    }
    send(method, args) {
        return new Promise((resolve, reject) => {
            const id = crypto_1.default.randomUUID();
            this.pendingRequests[id] = { resolve, reject, method, args };
            this.process.send?.({ type: 'rpc_request', id, method, args });
        });
    }
    sendNoWait(method, args) {
        this.process.send?.({ type: 'rpc_request', method, args });
    }
    init() {
        this.process.on('message', (msg) => {
            if (msg.type === 'rpc_response') {
                const { id, result, error } = msg;
                const callbacks = this.pendingRequests[id];
                if (!callbacks) {
                    return;
                }
                else if (error) {
                    callbacks.reject(error);
                }
                else {
                    callbacks.resolve(result);
                }
                delete this.pendingRequests[id];
            }
        });
    }
}
exports.RpcSender = RpcSender;
