// motia/dist/esm/create/interactive.js
import colors from 'colors';
import inquirer from 'inquirer';
import { create } from './index';
const choices = {
    nodejs: 'Tutorial (TypeScript)',
    python: 'Tutorial (Python)',
    'starter-typescript': 'Starter (TypeScript)',
    'starter-javascript': 'Starter (JavaScript)',
    'starter-python': 'Starter (Python)',
};
export const createInteractive = async (args, context) => {
    context.log('welcome', (message) => message.append(`\n🚀 ${colors.bold(args.plugin ? 'Welcome to Motia Plugin Creator!' : 'Welcome to Motia Project Creator!')}`));
    const questions = [];
    let name = args.name;
    let template = args.template;
    if (args.plugin) {
        if (!args.name) {
            context.log('failed', (message) => message
                .tag('failed')
                .append(`Project name is required: ${colors.bold('motia create --plugin [project-name]')}\n`));
            return;
        }
        return create({
            projectName: args.name,
            template: 'plugin',
            cursorEnabled: false,
            context,
        });
    }
    else if (!args.template) {
        questions.push({
            type: 'list',
            name: 'template',
            message: 'What template do you want to use? (Use arrow keys)',
            choices: Object.keys(choices).map((key) => ({
                name: choices[key],
                value: key,
            })),
        });
    }
    if (!args.name) {
        questions.push({
            type: 'input',
            name: 'projectName',
            message: 'Project name (leave blank to use current folder):',
            validate: (input) => {
                if (input && input.trim().length > 0) {
                    if (!/^[a-zA-Z0-9][a-zA-Z0-9-_]*$/.test(input.trim())) {
                        return 'Project name must start with a letter or number and contain only letters, numbers, hyphens, and underscores';
                    }
                }
                return true;
            },
            filter: (input) => input.trim(),
        });
    }
    if (questions.length > 0) {
        const answers = await inquirer.prompt(questions);
        name = args.name || answers.projectName;
        template = args.template || answers.template;
    }
    context.log('creating', (message) => message.append('\n🔨 Creating your Motia project...\n'));
    await create({
        projectName: name || '.',
        template: template || 'nodejs',
        cursorEnabled: true, // Default to true for cursor rules
        context,
    });
    process.exit(0);
};
