// motia/dist/esm/start.js
import { RedisStateAdapter } from '@motiadev/adapter-redis-state';
import { RedisStreamAdapterManager } from '@motiadev/adapter-redis-streams';
import { createServer, DefaultCronAdapter, DefaultQueueEventAdapter } from '@motiadev/core';
import path from 'path';
import { workbenchBase } from './constants';
import { generateLockedData, getStepFiles } from './generate-locked-data';
import { loadMotiaConfig } from './load-motia-config';
import { processPlugins } from './plugins';
import { instanceRedisMemoryServer, stopRedisMemoryServer } from './redis-memory-manager';
import { activatePythonVenv } from './utils/activate-python-env';
import { version } from './version';
require('ts-node').register({
    transpileOnly: true,
    compilerOptions: { module: 'commonjs' },
});
export const start = async (port, hostname, disableVerbose, motiaFileStorageDir) => {
    const baseDir = process.cwd();
    const isVerbose = !disableVerbose;
    const stepFiles = getStepFiles(baseDir);
    const hasPythonFiles = stepFiles.some((file) => file.endsWith('.py'));
    if (hasPythonFiles) {
        console.log('âš™ï¸ Activating Python environment...');
        activatePythonVenv({ baseDir, isVerbose });
    }
    const motiaFileStoragePath = motiaFileStorageDir || '.motia';
    const dotMotia = path.join(baseDir, motiaFileStoragePath);
    const appConfig = await loadMotiaConfig(baseDir);
    const redisClient = await instanceRedisMemoryServer(dotMotia);
    const adapters = {
        eventAdapter: appConfig.adapters?.events || new DefaultQueueEventAdapter(),
        cronAdapter: appConfig.adapters?.cron || new DefaultCronAdapter(),
        streamAdapter: appConfig.adapters?.streams || new RedisStreamAdapterManager(redisClient),
    };
    const lockedData = await generateLockedData({
        projectDir: baseDir,
        streamAdapter: adapters.streamAdapter,
        redisClient,
    });
    const state = appConfig.adapters?.state || new RedisStateAdapter(redisClient);
    const config = { isVerbose, isDev: false, version };
    const motiaServer = createServer(lockedData, state, config, adapters, appConfig.app);
    const plugins = await processPlugins(motiaServer);
    if (!process.env.MOTIA_DOCKER_DISABLE_WORKBENCH) {
        const { applyMiddleware } = require('@motiadev/workbench/dist/middleware');
        await applyMiddleware({
            app: motiaServer.app,
            port,
            workbenchBase,
            plugins: plugins.flatMap((item) => item.workbench),
        });
    }
    motiaServer.server.listen(port, hostname);
    console.log('ðŸš€ Server ready and listening on port', port);
    console.log(`ðŸ”— Open http://${hostname}:${port}${workbenchBase} to open workbench ðŸ› ï¸`);
    process.on('SIGTERM', async () => {
        motiaServer.server.close();
        await stopRedisMemoryServer();
        process.exit(0);
    });
    process.on('SIGINT', async () => {
        motiaServer.server.close();
        await stopRedisMemoryServer();
        process.exit(0);
    });
};
