// @motiadev/adapter-redis-streams/dist/redis-stream-adapter-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisStreamAdapterManager = void 0;
const redis_1 = require("redis");
const redis_stream_adapter_1 = require("./redis-stream-adapter");
function isRedisClient(input) {
    return typeof input === 'object' && 'isOpen' in input && 'connect' in input;
}
class RedisStreamAdapterManager {
    constructor(redisConnection, options) {
        this.connected = false;
        this.config = {
            keyPrefix: options?.keyPrefix || 'motia:stream:',
            socketKeepAlive: options?.socketKeepAlive ?? true,
        };
        if (isRedisClient(redisConnection)) {
            this.client = redisConnection;
            this.isExternalClient = true;
            this.connected = this.client.isOpen;
        }
        else {
            const config = {
                ...redisConnection,
                socket: {
                    ...(redisConnection.socket || {}),
                    keepAlive: this.config.socketKeepAlive,
                    noDelay: true,
                },
            };
            this.isExternalClient = false;
            this.client = (0, redis_1.createClient)(config);
            this.client.on('error', (err) => {
                console.error('[Redis Stream Manager] Client error:', err);
            });
            this.client.on('connect', () => {
                this.connected = true;
            });
            this.client.on('disconnect', () => {
                console.warn('[Redis Stream Manager] Disconnected');
                this.connected = false;
            });
            this.connect();
        }
    }
    async connect() {
        if (!this.connected && !this.isExternalClient) {
            try {
                await this.client.connect();
            }
            catch (error) {
                console.error('[Redis Stream Manager] Failed to connect:', error);
                throw error;
            }
        }
    }
    createStream(streamName) {
        return new redis_stream_adapter_1.RedisStreamAdapter(streamName, this.config, this.client);
    }
    getClient() {
        return this.client;
    }
    async shutdown() {
        if (!this.isExternalClient && this.client.isOpen) {
            await this.client.quit();
        }
    }
}
exports.RedisStreamAdapterManager = RedisStreamAdapterManager;
