// @motiadev/workbench/dist/middleware.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyMiddleware = void 0;
const plugin_react_1 = __importDefault(require("@vitejs/plugin-react"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const vite_1 = require("vite");
const motia_plugin_1 = __importDefault(require("./motia-plugin"));
const workbenchBasePlugin = (workbenchBase) => {
    return {
        name: 'html-transform',
        transformIndexHtml: (html) => {
            return html.replace('</head>', `<script>const workbenchBase = ${JSON.stringify(workbenchBase)};</script></head>`);
        },
    };
};
const processCwdPlugin = () => {
    return {
        name: 'html-transform',
        transformIndexHtml: (html) => {
            // Normalize path for cross-platform compatibility
            const cwd = process.cwd().replace(/\\/g, '/');
            return html.replace('</head>', `<script>const processCwd = "${cwd}";</script></head>`);
        },
    };
};
const reoPlugin = () => {
    return {
        name: 'html-transform',
        transformIndexHtml(html) {
            const isAnalyticsEnabled = process.env.MOTIA_ANALYTICS_DISABLED !== 'true';
            if (!isAnalyticsEnabled) {
                return html;
            }
            // inject before </head>
            return html.replace('</head>', `
        <script type="text/javascript">
          !function(){var e,t,n;e="d8f0ce9cae8ae64",t=function(){Reo.init({clientID:"d8f0ce9cae8ae64", source: "internal"})},(n=document.createElement("script")).src="https://static.reo.dev/"+e+"/reo.js",n.defer=!0,n.onload=t,document.head.appendChild(n)}();
        </script>
    </head>`);
        },
    };
};
const applyMiddleware = async ({ app, port, workbenchBase, plugins }) => {
    const vite = await (0, vite_1.createServer)({
        appType: 'spa',
        root: __dirname,
        base: workbenchBase,
        server: {
            middlewareMode: true,
            allowedHosts: true,
            host: true,
            hmr: { port: 21678 + port },
            fs: {
                allow: [
                    __dirname, // workbench root
                    path_1.default.join(process.cwd(), './steps'), // steps directory
                    path_1.default.join(process.cwd(), './src'), // src directory
                    path_1.default.join(process.cwd(), './tutorial'), // tutorial directory
                    path_1.default.join(process.cwd(), './node_modules'), // node_modules directory
                    path_1.default.join(__dirname, './node_modules'), // node_modules directory
                ],
            },
        },
        resolve: {
            alias: {
                '@': path_1.default.resolve(__dirname, './src'),
                '@/assets': path_1.default.resolve(__dirname, './src/assets'),
                'lucide-react/dynamic': 'lucide-react/dynamic.mjs',
                'lucide-react': 'lucide-react/dist/cjs/lucide-react.js',
            },
        },
        plugins: [
            (0, plugin_react_1.default)(),
            processCwdPlugin(),
            reoPlugin(),
            (0, motia_plugin_1.default)(plugins),
            workbenchBasePlugin(workbenchBase),
        ],
        assetsInclude: ['**/*.png', '**/*.jpg', '**/*.jpeg', '**/*.gif', '**/*.svg', '**/*.ico', '**/*.webp', '**/*.avif'],
    });
    app.use(workbenchBase, vite.middlewares);
    app.use(`${workbenchBase}/*`, async (req, res, next) => {
        const url = req.originalUrl;
        try {
            const index = fs_1.default.readFileSync(path_1.default.resolve(__dirname, 'index.html'), 'utf-8');
            const html = await vite.transformIndexHtml(url, index);
            res.status(200).set({ 'Content-Type': 'text/html' }).end(html);
        }
        catch (e) {
            next(e);
        }
    });
};
exports.applyMiddleware = applyMiddleware;
