// @motiadev/stream-client/dist/cjs/src/stream-subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamSubscription = void 0;
class StreamSubscription {
    constructor(sub, state) {
        this.customEventListeners = new Map();
        this.closeListeners = new Set();
        this.onChangeListeners = new Set();
        this.sub = sub;
        this.state = state;
    }
    onEventReceived(event) {
        const customEventListeners = this.customEventListeners.get(event.type);
        if (customEventListeners) {
            const eventData = event.data;
            customEventListeners.forEach((listener) => listener(eventData));
        }
    }
    /**
     * Add a custom event listener. This listener will be called whenever the custom event is received.
     */
    onEvent(type, listener) {
        const listeners = this.customEventListeners.get(type) || [];
        this.customEventListeners.set(type, [...listeners, listener]);
    }
    /**
     * Remove a custom event listener.
     */
    offEvent(type, listener) {
        const listeners = this.customEventListeners.get(type) || [];
        this.customEventListeners.set(type, listeners.filter((l) => l !== listener));
    }
    onClose(listener) {
        this.closeListeners.add(listener);
    }
    close() {
        this.closeListeners.forEach((listener) => listener());
        this.closeListeners.clear();
    }
    /**
     * Add a change listener. This listener will be called whenever the state of the group changes.
     */
    addChangeListener(listener) {
        this.onChangeListeners.add(listener);
    }
    /**
     * Remove a change listener.
     */
    removeChangeListener(listener) {
        this.onChangeListeners.delete(listener);
    }
    /**
     * Get the current state of the group.
     */
    getState() {
        return this.state;
    }
    setState(state) {
        this.state = state;
        this.onChangeListeners.forEach((listener) => listener(state));
    }
}
exports.StreamSubscription = StreamSubscription;
